// @verdaccio/middleware/build/middlewares/web/utils/renderHTML.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = renderHTML;
var _debug = _interopRequireDefault(require("debug"));
var _lruCache = _interopRequireDefault(require("lru-cache"));
var _core = require("@verdaccio/core");
var _template = _interopRequireDefault(require("./template"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
// Cache for rendered HTML templates: max 500 entries, 1 hour TTL
const cache = new _lruCache.default({
  max: 500,
  ttl: 1000 * 60 * 60
});
const debug = (0, _debug.default)('verdaccio:web:render');
const defaultManifestFiles = {
  js: ['vendors.js', 'main.js'],
  ico: 'favicon.ico',
  css: []
};
function renderHTML(config, manifest, manifestFiles, options, res) {
  // @ts-ignore
  const needHtmlCache = [undefined, null].includes(config?.web?.html_cache) ? true : config?.web?.html_cache;
  const scriptsBodyBefore = config?.web?.scriptsBodyBefore || config?.web?.scriptsbodyBefore || [];
  const scriptsBodyAfter = config?.web?.scriptsBodyAfter || [];
  const metaScripts = config?.web?.metaScripts || [];
  let webPage;
  const cacheKey = `template:${JSON.stringify(options)}`;
  try {
    webPage = cache.get(cacheKey);
    if (!webPage) {
      webPage = (0, _template.default)({
        manifest: manifestFiles ?? defaultManifestFiles,
        options,
        scriptsBodyAfter,
        metaScripts,
        scriptsBodyBefore
      }, manifest);
      if (needHtmlCache) {
        cache.set(cacheKey, webPage);
        debug('set template cache');
      }
    } else {
      debug('reuse template cache');
    }
  } catch (error) {
    throw new Error(`theme could not be load, stack ${error.stack}`);
  }
  res.setHeader('Content-Type', _core.HEADERS.TEXT_HTML);
  res.send(webPage);
  debug('web rendered');
}
