// @verdaccio/file-locking/build/legacy/utils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.statfile = exports.statDir = exports.lockfile = void 0;
var _lockfile = _interopRequireDefault(require("lockfile"));
var _nodeFs = _interopRequireDefault(require("node:fs"));
var _nodePath = _interopRequireDefault(require("node:path"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { "default": e }; }
var statDir = exports.statDir = function statDir(name) {
  return new Promise(function (resolve, reject) {
    // test to see if the directory exists
    var dirPath = _nodePath["default"].dirname(name);
    _nodeFs["default"].stat(dirPath, function (err, stats) {
      if (err) {
        return reject(err);
      } else if (!stats.isDirectory()) {
        return resolve(new Error("".concat(_nodePath["default"].dirname(name), " is not a directory")));
      } else {
        return resolve(null);
      }
    });
  });
};
var statfile = exports.statfile = function statfile(name) {
  return new Promise(function (resolve, reject) {
    // test to see if the directory exists
    _nodeFs["default"].stat(name, function (err, stats) {
      if (err) {
        return reject(err);
      } else if (!stats.isFile()) {
        return resolve(new Error("".concat(_nodePath["default"].dirname(name), " is not a file")));
      } else {
        return resolve(null);
      }
    });
  });
};
var lockfile = exports.lockfile = function lockfile(name) {
  return new Promise(function (resolve) {
    var lockOpts = {
      // time (ms) to wait when checking for stale locks
      wait: 1000,
      // how often (ms) to re-check stale locks
      pollPeriod: 100,
      // locks are considered stale after 5 minutes
      stale: 5 * 60 * 1000,
      // number of times to attempt to create a lock
      retries: 100,
      // time (ms) between tries
      retryWait: 100
    };
    var lockFileName = "".concat(name, ".lock");
    _lockfile["default"].lock(lockFileName, lockOpts, function () {
      resolve();
    });
  });
};
