// @verdaccio/auth/build/utils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.allow_action = allow_action;
exports.buildUser = buildUser;
exports.convertPayloadToBase64 = convertPayloadToBase64;
exports.expireReasons = void 0;
exports.getApiToken = getApiToken;
exports.getDefaultPluginMethods = getDefaultPluginMethods;
exports.getMiddlewareCredentials = getMiddlewareCredentials;
exports.handleSpecialUnpublish = handleSpecialUnpublish;
exports.isAESLegacy = isAESLegacy;
exports.isAuthHeaderValid = isAuthHeaderValid;
exports.parseAESCredentials = parseAESCredentials;
exports.parseAuthTokenHeader = parseAuthTokenHeader;
exports.verifyJWTPayload = verifyJWTPayload;
var _debug = _interopRequireDefault(require("debug"));
var _lodash = _interopRequireDefault(require("lodash"));
var _config = require("@verdaccio/config");
var _core = require("@verdaccio/core");
var _signature = require("@verdaccio/signature");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { "default": e }; }
function _regenerator() { /*! regenerator-runtime -- Copyright (c) 2014-present, Facebook, Inc. -- license (MIT): https://github.com/babel/babel/blob/main/packages/babel-helpers/LICENSE */ var e, t, r = "function" == typeof Symbol ? Symbol : {}, n = r.iterator || "@@iterator", o = r.toStringTag || "@@toStringTag"; function i(r, n, o, i) { var c = n && n.prototype instanceof Generator ? n : Generator, u = Object.create(c.prototype); return _regeneratorDefine2(u, "_invoke", function (r, n, o) { var i, c, u, f = 0, p = o || [], y = !1, G = { p: 0, n: 0, v: e, a: d, f: d.bind(e, 4), d: function d(t, r) { return i = t, c = 0, u = e, G.n = r, a; } }; function d(r, n) { for (c = r, u = n, t = 0; !y && f && !o && t < p.length; t++) { var o, i = p[t], d = G.p, l = i[2]; r > 3 ? (o = l === n) && (u = i[(c = i[4]) ? 5 : (c = 3, 3)], i[4] = i[5] = e) : i[0] <= d && ((o = r < 2 && d < i[1]) ? (c = 0, G.v = n, G.n = i[1]) : d < l && (o = r < 3 || i[0] > n || n > l) && (i[4] = r, i[5] = n, G.n = l, c = 0)); } if (o || r > 1) return a; throw y = !0, n; } return function (o, p, l) { if (f > 1) throw TypeError("Generator is already running"); for (y && 1 === p && d(p, l), c = p, u = l; (t = c < 2 ? e : u) || !y;) { i || (c ? c < 3 ? (c > 1 && (G.n = -1), d(c, u)) : G.n = u : G.v = u); try { if (f = 2, i) { if (c || (o = "next"), t = i[o]) { if (!(t = t.call(i, u))) throw TypeError("iterator result is not an object"); if (!t.done) return t; u = t.value, c < 2 && (c = 0); } else 1 === c && (t = i["return"]) && t.call(i), c < 2 && (u = TypeError("The iterator does not provide a '" + o + "' method"), c = 1); i = e; } else if ((t = (y = G.n < 0) ? u : r.call(n, G)) !== a) break; } catch (t) { i = e, c = 1, u = t; } finally { f = 1; } } return { value: t, done: y }; }; }(r, o, i), !0), u; } var a = {}; function Generator() {} function GeneratorFunction() {} function GeneratorFunctionPrototype() {} t = Object.getPrototypeOf; var c = [][n] ? t(t([][n]())) : (_regeneratorDefine2(t = {}, n, function () { return this; }), t), u = GeneratorFunctionPrototype.prototype = Generator.prototype = Object.create(c); function f(e) { return Object.setPrototypeOf ? Object.setPrototypeOf(e, GeneratorFunctionPrototype) : (e.__proto__ = GeneratorFunctionPrototype, _regeneratorDefine2(e, o, "GeneratorFunction")), e.prototype = Object.create(u), e; } return GeneratorFunction.prototype = GeneratorFunctionPrototype, _regeneratorDefine2(u, "constructor", GeneratorFunctionPrototype), _regeneratorDefine2(GeneratorFunctionPrototype, "constructor", GeneratorFunction), GeneratorFunction.displayName = "GeneratorFunction", _regeneratorDefine2(GeneratorFunctionPrototype, o, "GeneratorFunction"), _regeneratorDefine2(u), _regeneratorDefine2(u, o, "Generator"), _regeneratorDefine2(u, n, function () { return this; }), _regeneratorDefine2(u, "toString", function () { return "[object Generator]"; }), (_regenerator = function _regenerator() { return { w: i, m: f }; })(); }
function _regeneratorDefine2(e, r, n, t) { var i = Object.defineProperty; try { i({}, "", {}); } catch (e) { i = 0; } _regeneratorDefine2 = function _regeneratorDefine(e, r, n, t) { function o(r, n) { _regeneratorDefine2(e, r, function (e) { return this._invoke(r, n, e); }); } r ? i ? i(e, r, { value: n, enumerable: !t, configurable: !t, writable: !t }) : e[r] = n : (o("next", 0), o("throw", 1), o("return", 2)); }, _regeneratorDefine2(e, r, n, t); }
function asyncGeneratorStep(n, t, e, r, o, a, c) { try { var i = n[a](c), u = i.value; } catch (n) { return void e(n); } i.done ? t(u) : Promise.resolve(u).then(r, o); }
function _asyncToGenerator(n) { return function () { var t = this, e = arguments; return new Promise(function (r, o) { var a = n.apply(t, e); function _next(n) { asyncGeneratorStep(a, r, o, _next, _throw, "next", n); } function _throw(n) { asyncGeneratorStep(a, r, o, _next, _throw, "throw", n); } _next(void 0); }); }; }
function _slicedToArray(r, e) { return _arrayWithHoles(r) || _iterableToArrayLimit(r, e) || _unsupportedIterableToArray(r, e) || _nonIterableRest(); }
function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _unsupportedIterableToArray(r, a) { if (r) { if ("string" == typeof r) return _arrayLikeToArray(r, a); var t = {}.toString.call(r).slice(8, -1); return "Object" === t && r.constructor && (t = r.constructor.name), "Map" === t || "Set" === t ? Array.from(r) : "Arguments" === t || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(t) ? _arrayLikeToArray(r, a) : void 0; } }
function _arrayLikeToArray(r, a) { (null == a || a > r.length) && (a = r.length); for (var e = 0, n = Array(a); e < a; e++) n[e] = r[e]; return n; }
function _iterableToArrayLimit(r, l) { var t = null == r ? null : "undefined" != typeof Symbol && r[Symbol.iterator] || r["@@iterator"]; if (null != t) { var e, n, i, u, a = [], f = !0, o = !1; try { if (i = (t = t.call(r)).next, 0 === l) { if (Object(t) !== t) return; f = !1; } else for (; !(f = (e = i.call(t)).done) && (a.push(e.value), a.length !== l); f = !0); } catch (r) { o = !0, n = r; } finally { try { if (!f && null != t["return"] && (u = t["return"](), Object(u) !== u)) return; } finally { if (o) throw n; } } return a; } }
function _arrayWithHoles(r) { if (Array.isArray(r)) return r; }
var debug = (0, _debug["default"])('verdaccio:auth:utils');

/**
 * Split authentication header eg: Bearer [secret_token]
 * @param authorizationHeader auth token
 */
function parseAuthTokenHeader(authorizationHeader) {
  var parts = authorizationHeader.split(' ');
  var _parts = _slicedToArray(parts, 2),
    scheme = _parts[0],
    token = _parts[1];
  return {
    scheme: scheme,
    token: token
  };
}
function parseAESCredentials(authorizationHeader, secret) {
  debug('parseAESCredentials init');
  var _parseAuthTokenHeader = parseAuthTokenHeader(authorizationHeader),
    scheme = _parseAuthTokenHeader.scheme,
    token = _parseAuthTokenHeader.token;

  // basic is deprecated and should not be enforced
  // basic is currently being used for functional test
  if (scheme.toUpperCase() === _core.TOKEN_BASIC.toUpperCase()) {
    debug('legacy header basic');
    var credentials = convertPayloadToBase64(token).toString();
    return credentials;
  } else if (scheme.toUpperCase() === _core.TOKEN_BEARER.toUpperCase()) {
    debug('legacy header bearer');
    debug('secret length %o', secret.length);
    var isLegacyUnsecure = secret.length > 32;
    debug('is legacy unsecure %o', isLegacyUnsecure);
    if (isLegacyUnsecure) {
      debug('legacy unsecure enabled');
      return (0, _signature.aesDecryptDeprecated)(convertPayloadToBase64(token), secret).toString('utf-8');
    } else {
      debug('legacy secure enabled');
      return (0, _signature.aesDecrypt)(token.toString(), secret);
    }
  }
}
function getMiddlewareCredentials(security, secretKey, authorizationHeader) {
  debug('getMiddlewareCredentials init');
  // comment out for debugging purposes
  if (isAESLegacy(security)) {
    debug('is legacy');
    var credentials = parseAESCredentials(authorizationHeader, secretKey);
    if (!credentials) {
      debug('parse legacy credentials failed');
      return;
    }
    var parsedCredentials = (0, _signature.parseBasicPayload)(credentials);
    if (!parsedCredentials) {
      debug('parse legacy basic payload credentials failed');
      return;
    }
    return parsedCredentials;
  }
  var _parseAuthTokenHeader2 = parseAuthTokenHeader(authorizationHeader),
    scheme = _parseAuthTokenHeader2.scheme,
    token = _parseAuthTokenHeader2.token;
  debug('is jwt');
  if (_lodash["default"].isString(token) && scheme.toUpperCase() === _core.TOKEN_BEARER.toUpperCase()) {
    return verifyJWTPayload(token, secretKey, security);
  }
}
function isAESLegacy(security) {
  var _security$api = security.api,
    legacy = _security$api.legacy,
    jwt = _security$api.jwt;
  return _lodash["default"].isNil(legacy) === false && _lodash["default"].isNil(jwt) && legacy === true;
}
function getApiToken(_x, _x2, _x3, _x4) {
  return _getApiToken.apply(this, arguments);
}
function _getApiToken() {
  _getApiToken = _asyncToGenerator(/*#__PURE__*/_regenerator().m(function _callee(auth, config, remoteUser, aesPassword) {
    var security, jwt;
    return _regenerator().w(function (_context) {
      while (1) switch (_context.n) {
        case 0:
          debug('get api token');
          security = config.security;
          if (!isAESLegacy(security)) {
            _context.n = 2;
            break;
          }
          debug('security legacy enabled');
          // fallback all goes to AES encryption
          _context.n = 1;
          return new Promise(function (resolve) {
            resolve(auth.aesEncrypt(buildUser(remoteUser.name, aesPassword)));
          });
        case 1:
          return _context.a(2, _context.v);
        case 2:
          jwt = security.api.jwt;
          if (!(jwt !== null && jwt !== void 0 && jwt.sign)) {
            _context.n = 4;
            break;
          }
          _context.n = 3;
          return auth.jwtEncrypt(remoteUser, jwt.sign);
        case 3:
          return _context.a(2, _context.v);
        case 4:
          _context.n = 5;
          return new Promise(function (resolve) {
            resolve(auth.aesEncrypt(buildUser(remoteUser.name, aesPassword)));
          });
        case 5:
          return _context.a(2, _context.v);
      }
    }, _callee);
  }));
  return _getApiToken.apply(this, arguments);
}
var expireReasons = exports.expireReasons = ['JsonWebTokenError', 'TokenExpiredError'];
function verifyJWTPayload(token, secret, security) {
  try {
    var _security$api2;
    var payload = (0, _signature.verifyPayload)(token, secret, security === null || security === void 0 || (_security$api2 = security.api) === null || _security$api2 === void 0 || (_security$api2 = _security$api2.jwt) === null || _security$api2 === void 0 ? void 0 : _security$api2.verify);
    return payload;
  } catch (error) {
    // #168 this check should be removed as soon AES encrypt is removed.
    if (expireReasons.includes(error.name)) {
      // it might be possible the jwt configuration is enabled and
      // old tokens fails still remains in usage, thus
      // we return an anonymous user to force log in.
      return (0, _config.createAnonymousRemoteUser)();
    }
    throw _core.errorUtils.getCode(_core.HTTP_STATUS.UNAUTHORIZED, error.message);
  }
}
function isAuthHeaderValid(authorization) {
  return authorization.split(' ').length === 2;
}

/**
 * Return a default configuration for authentication if none is provided.
 * @param logger {Logger}
 * @returns object of default implementations.
 */
function getDefaultPluginMethods(logger) {
  return {
    authenticate: function authenticate(_user, _password, cb) {
      debug('triggered default authenticate method');
      cb(_core.errorUtils.getForbidden(_core.API_ERROR.BAD_USERNAME_PASSWORD));
    },
    adduser: function adduser(_user, _password, cb) {
      debug('triggered default adduser method');
      // since adduser is not implemented but optional, continue without error
      // this assumes that the user is added by an external system
      cb(null, true);
    },
    // @ts-ignore
    allow_access: allow_action('access', logger),
    // @ts-ignore
    allow_publish: allow_action('publish', logger),
    allow_unpublish: handleSpecialUnpublish(logger)
  };
}
function allow_action(action, logger) {
  return function allowActionCallback(user, pkg, callback) {
    logger.trace({
      remote: user.name
    }, "[auth/allow_action]: user: @{remote}");
    var name = user.name,
      groups = user.groups;
    debug('allow_action "%s": groups %s', action, groups);
    var groupAccess = pkg[action];
    debug('allow_action "%s": groupAccess %s', action, groupAccess);
    var hasPermission = groupAccess.some(function (group) {
      return name === group || groups.includes(group);
    });
    debug('package "%s" has permission "%s"', name, hasPermission);
    logger.trace({
      pkgName: pkg.name,
      hasPermission: hasPermission,
      remote: user.name,
      groupAccess: groupAccess
    }, "[auth/allow_action]: hasPermission? @{hasPermission} for user: @{remote}, package: @{pkgName}");
    if (hasPermission) {
      logger.trace({
        remote: user.name
      }, "auth/allow_action: access granted to: @{remote}");
      return callback(null, true);
    }
    if (name) {
      callback(_core.errorUtils.getForbidden("user ".concat(name, " is not allowed to ").concat(action, " package ").concat(pkg.name)));
    } else {
      callback(_core.errorUtils.getUnauthorized("authorization required to ".concat(action, " package ").concat(pkg.name)));
    }
  };
}

/**
 *
 */
function handleSpecialUnpublish(logger) {
  return function (user, pkg, callback) {
    var action = 'unpublish';
    // verify whether the unpublish prop has been defined
    var isUnpublishMissing = !pkg[action];
    debug('is unpublish method missing ? %s', isUnpublishMissing);
    var hasGroups = isUnpublishMissing ? false : pkg[action].length > 0;
    logger.trace({
      user: user.name,
      name: pkg.name,
      hasGroups: hasGroups
    }, "fallback unpublish for @{name} has groups: @{hasGroups} for @{user}");
    if (isUnpublishMissing || hasGroups === false) {
      return callback(null, undefined);
    }
    logger.trace({
      user: user.name,
      name: pkg.name,
      action: action,
      hasGroups: hasGroups
    }, "allow_action for @{action} for @{name} has groups: @{hasGroups} for @{user}");
    return allow_action(action, logger)(user, pkg, callback);
  };
}
function buildUser(name, password) {
  return String("".concat(name, ":").concat(password));
}
function convertPayloadToBase64(payload) {
  return Buffer.from(payload, 'base64');
}
