// verdaccio/build/lib/cli/commands/init.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.InitCommand = exports.DEFAULT_PROCESS_NAME = void 0;
var _clipanion = require("clipanion");
var _path = _interopRequireDefault(require("path"));
var _config = require("@verdaccio/config");
var _bootstrap = require("../../bootstrap");
var _logger = require("../../logger");
var _runServer = require("../../run-server");
var _utils = require("../../utils");
var _utils2 = require("../utils");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const pkgVersion = "6.7.1" || 'dev';
const pkgName = 'verdaccio';
const DEFAULT_PROCESS_NAME = exports.DEFAULT_PROCESS_NAME = 'verdaccio';
class InitCommand extends _clipanion.Command {
  static paths = [_clipanion.Command.Default];
  listen = _clipanion.Option.String('-l,--listen', {
    description: 'host:port number to listen on (default: localhost:4873)'
  });
  static usage = _clipanion.Command.Usage({
    description: `launch the server`,
    details: `
      This start the registry in the default port.

      When used without arguments, it:

      - bootstrap the server at the port  \`4873\`

      The optional arguments are:

      - \`--listen\` to switch the default server port,
      - \`--config\` to define a different configuration path location,

    `,
    examples: [[`Runs the server with the default configuration`, `verdaccio`], [`Runs the server in the port 5000`, `verdaccio --listen 5000`], [`Runs the server by using a different absolute location of the configuration file`, `verdaccio --config /home/user/verdaccio/config.yaml`]]
  });
  config = _clipanion.Option.String('-c,--config', {
    description: 'use this configuration file (default: ./config.yaml)'
  });
  async execute() {
    let configPathLocation;
    try {
      configPathLocation = (0, _config.findConfigFile)(this.config);
      const configParsed = (0, _utils.parseConfigFile)(configPathLocation);
      if (!configParsed.self_path) {
        configParsed.self_path = _path.default.resolve(configPathLocation);
        // compatibility with 6.x plugins
        configParsed.configPath = configParsed.self_path;
      }
      if (!configParsed.https) {
        configParsed.https = {
          enable: false
        };
      }
      process.title = configParsed.web && configParsed.web.title || 'verdaccio';
      (0, _utils.initLogger)(configParsed);
      if (!(0, _utils2.isVersionRecommended)(process.version)) {
        _logger.logger.warn({
          current: process.version,
          recommended: _utils2.RECOMMENDED_NODE_VERSION
        }, 'you are using Node.js @{current}, Verdaccio recommends Node.js v@{recommended} or higher, please consider upgrading your Node.js distribution');
      }
      _logger.logger.info({
        file: configPathLocation
      }, 'config file  - @{file}');
      const webServer = await (0, _runServer.runServer)(configParsed, {
        listenArg: this.listen
      });
      const listen = this.listen ?? configParsed.listen;
      const addr = (0, _config.getListenAddress)(listen, _logger.logger);
      (0, _bootstrap.listenDefaultCallback)(webServer, addr, pkgName, pkgVersion);
    } catch (err) {
      console.error(`cannot open config file ${configPathLocation}: ${err.stack}`);
      // @ts-expect-error
      if (typeof _logger.logger?.logger?.fatal === 'function') {
        _logger.logger.fatal({
          file: configPathLocation,
          err: err
        }, 'cannot open config file @{file}: @{!err.message}');
      } else {
        console.error(`cannot open config file ${configPathLocation}: ${!err.message}`);
      }
      process.exit(1);
    }
  }
}
exports.InitCommand = InitCommand;
