// verdaccio-htpasswd/build/htpasswd.js
"use strict";

function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = exports.DEFAULT_SLOW_VERIFY_MS = void 0;
var _debug = _interopRequireDefault(require("debug"));
var _nodeFs = require("node:fs");
var _nodePath = require("node:path");
var _core = require("@verdaccio/core");
var _fileLocking = require("@verdaccio/file-locking");
var _utils = require("./utils");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { "default": e }; }
function _regenerator() { /*! regenerator-runtime -- Copyright (c) 2014-present, Facebook, Inc. -- license (MIT): https://github.com/babel/babel/blob/main/packages/babel-helpers/LICENSE */ var e, t, r = "function" == typeof Symbol ? Symbol : {}, n = r.iterator || "@@iterator", o = r.toStringTag || "@@toStringTag"; function i(r, n, o, i) { var c = n && n.prototype instanceof Generator ? n : Generator, u = Object.create(c.prototype); return _regeneratorDefine2(u, "_invoke", function (r, n, o) { var i, c, u, f = 0, p = o || [], y = !1, G = { p: 0, n: 0, v: e, a: d, f: d.bind(e, 4), d: function d(t, r) { return i = t, c = 0, u = e, G.n = r, a; } }; function d(r, n) { for (c = r, u = n, t = 0; !y && f && !o && t < p.length; t++) { var o, i = p[t], d = G.p, l = i[2]; r > 3 ? (o = l === n) && (u = i[(c = i[4]) ? 5 : (c = 3, 3)], i[4] = i[5] = e) : i[0] <= d && ((o = r < 2 && d < i[1]) ? (c = 0, G.v = n, G.n = i[1]) : d < l && (o = r < 3 || i[0] > n || n > l) && (i[4] = r, i[5] = n, G.n = l, c = 0)); } if (o || r > 1) return a; throw y = !0, n; } return function (o, p, l) { if (f > 1) throw TypeError("Generator is already running"); for (y && 1 === p && d(p, l), c = p, u = l; (t = c < 2 ? e : u) || !y;) { i || (c ? c < 3 ? (c > 1 && (G.n = -1), d(c, u)) : G.n = u : G.v = u); try { if (f = 2, i) { if (c || (o = "next"), t = i[o]) { if (!(t = t.call(i, u))) throw TypeError("iterator result is not an object"); if (!t.done) return t; u = t.value, c < 2 && (c = 0); } else 1 === c && (t = i["return"]) && t.call(i), c < 2 && (u = TypeError("The iterator does not provide a '" + o + "' method"), c = 1); i = e; } else if ((t = (y = G.n < 0) ? u : r.call(n, G)) !== a) break; } catch (t) { i = e, c = 1, u = t; } finally { f = 1; } } return { value: t, done: y }; }; }(r, o, i), !0), u; } var a = {}; function Generator() {} function GeneratorFunction() {} function GeneratorFunctionPrototype() {} t = Object.getPrototypeOf; var c = [][n] ? t(t([][n]())) : (_regeneratorDefine2(t = {}, n, function () { return this; }), t), u = GeneratorFunctionPrototype.prototype = Generator.prototype = Object.create(c); function f(e) { return Object.setPrototypeOf ? Object.setPrototypeOf(e, GeneratorFunctionPrototype) : (e.__proto__ = GeneratorFunctionPrototype, _regeneratorDefine2(e, o, "GeneratorFunction")), e.prototype = Object.create(u), e; } return GeneratorFunction.prototype = GeneratorFunctionPrototype, _regeneratorDefine2(u, "constructor", GeneratorFunctionPrototype), _regeneratorDefine2(GeneratorFunctionPrototype, "constructor", GeneratorFunction), GeneratorFunction.displayName = "GeneratorFunction", _regeneratorDefine2(GeneratorFunctionPrototype, o, "GeneratorFunction"), _regeneratorDefine2(u), _regeneratorDefine2(u, o, "Generator"), _regeneratorDefine2(u, n, function () { return this; }), _regeneratorDefine2(u, "toString", function () { return "[object Generator]"; }), (_regenerator = function _regenerator() { return { w: i, m: f }; })(); }
function _regeneratorDefine2(e, r, n, t) { var i = Object.defineProperty; try { i({}, "", {}); } catch (e) { i = 0; } _regeneratorDefine2 = function _regeneratorDefine(e, r, n, t) { function o(r, n) { _regeneratorDefine2(e, r, function (e) { return this._invoke(r, n, e); }); } r ? i ? i(e, r, { value: n, enumerable: !t, configurable: !t, writable: !t }) : e[r] = n : (o("next", 0), o("throw", 1), o("return", 2)); }, _regeneratorDefine2(e, r, n, t); }
function asyncGeneratorStep(n, t, e, r, o, a, c) { try { var i = n[a](c), u = i.value; } catch (n) { return void e(n); } i.done ? t(u) : Promise.resolve(u).then(r, o); }
function _asyncToGenerator(n) { return function () { var t = this, e = arguments; return new Promise(function (r, o) { var a = n.apply(t, e); function _next(n) { asyncGeneratorStep(a, r, o, _next, _throw, "next", n); } function _throw(n) { asyncGeneratorStep(a, r, o, _next, _throw, "throw", n); } _next(void 0); }); }; }
function _classCallCheck(a, n) { if (!(a instanceof n)) throw new TypeError("Cannot call a class as a function"); }
function _defineProperties(e, r) { for (var t = 0; t < r.length; t++) { var o = r[t]; o.enumerable = o.enumerable || !1, o.configurable = !0, "value" in o && (o.writable = !0), Object.defineProperty(e, _toPropertyKey(o.key), o); } }
function _createClass(e, r, t) { return r && _defineProperties(e.prototype, r), t && _defineProperties(e, t), Object.defineProperty(e, "prototype", { writable: !1 }), e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == _typeof(i) ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != _typeof(t) || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != _typeof(i)) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
function _callSuper(t, o, e) { return o = _getPrototypeOf(o), _possibleConstructorReturn(t, _isNativeReflectConstruct() ? Reflect.construct(o, e || [], _getPrototypeOf(t).constructor) : o.apply(t, e)); }
function _possibleConstructorReturn(t, e) { if (e && ("object" == _typeof(e) || "function" == typeof e)) return e; if (void 0 !== e) throw new TypeError("Derived constructors may only return object or undefined"); return _assertThisInitialized(t); }
function _assertThisInitialized(e) { if (void 0 === e) throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); return e; }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); }
function _getPrototypeOf(t) { return _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf.bind() : function (t) { return t.__proto__ || Object.getPrototypeOf(t); }, _getPrototypeOf(t); }
function _inherits(t, e) { if ("function" != typeof e && null !== e) throw new TypeError("Super expression must either be null or a function"); t.prototype = Object.create(e && e.prototype, { constructor: { value: t, writable: !0, configurable: !0 } }), Object.defineProperty(t, "prototype", { writable: !1 }), e && _setPrototypeOf(t, e); }
function _setPrototypeOf(t, e) { return _setPrototypeOf = Object.setPrototypeOf ? Object.setPrototypeOf.bind() : function (t, e) { return t.__proto__ = e, t; }, _setPrototypeOf(t, e); }
var debug = (0, _debug["default"])('verdaccio:plugin:htpasswd');
var DEFAULT_SLOW_VERIFY_MS = exports.DEFAULT_SLOW_VERIFY_MS = 200;

/**
 * HTPasswd - Verdaccio auth class
 */
var HTPasswd = exports["default"] = /*#__PURE__*/function (_ref) {
  /**
   *
   * @param {*} config htpasswd file
   * @param {object} options config.yaml in object from
   */

  // constructor
  function HTPasswd(config, options) {
    var _options$config, _options$config2;
    var _this;
    _classCallCheck(this, HTPasswd);
    _this = _callSuper(this, HTPasswd, [config, options]);
    _this.users = {};

    // verdaccio logger
    _this.logger = options.logger;

    // all this "verdaccio_config" stuff is for b/w compatibility only
    _this.maxUsers = config.max_users ? config.max_users : Infinity;
    var algorithm;
    var rounds;
    if (typeof config.algorithm === 'undefined') {
      algorithm = _core.constants.HtpasswdHashAlgorithm.bcrypt;
    } else if (_core.constants.HtpasswdHashAlgorithm[config.algorithm] !== undefined) {
      algorithm = _core.constants.HtpasswdHashAlgorithm[config.algorithm];
    } else {
      _this.logger.warn("The algorithm selected %s is invalid, switching to to default one \"bcrypt\", password validation can be affected", config.algorithm);
      algorithm = _core.constants.HtpasswdHashAlgorithm.bcrypt;
    }
    debug("password hash algorithm: ".concat(algorithm));
    if (algorithm === _core.constants.HtpasswdHashAlgorithm.bcrypt) {
      rounds = config.rounds || _utils.DEFAULT_BCRYPT_ROUNDS;
    } else if (config.rounds !== undefined) {
      _this.logger.warn({
        algo: algorithm
      }, 'Option "rounds" is not valid for "@{algo}" algorithm');
    }
    _this.hashConfig = {
      algorithm: algorithm,
      rounds: rounds
    };
    _this.lastTime = null;
    var file = config.file;
    debug('file: %s', file);
    if (!file) {
      throw new Error('should specify "file" in config');
    }
    debug('config path: %s', options === null || options === void 0 || (_options$config = options.config) === null || _options$config === void 0 ? void 0 : _options$config.configPath);
    _this.path = (0, _nodePath.resolve)((0, _nodePath.dirname)(options === null || options === void 0 || (_options$config2 = options.config) === null || _options$config2 === void 0 ? void 0 : _options$config2.configPath), file);
    _this.logger.info({
      file: _this.path
    }, 'using htpasswd file: @{file}');
    debug('htpasswd path:', _this.path);
    if (config.slow_verify_ms) {
      _this.logger.info({
        ms: config.slow_verify_ms
      }, 'slow_verify_ms enabled for @{ms}');
    }
    _this.slowVerifyMs = config.slow_verify_ms || DEFAULT_SLOW_VERIFY_MS;
    return _this;
  }

  /**
   * authenticate - Authenticate user.
   * @param {string} user
   * @param {string} password
   * @param {function} cb
   * @returns {void}
   */
  _inherits(HTPasswd, _ref);
  return _createClass(HTPasswd, [{
    key: "authenticate",
    value: function authenticate(user, password, cb) {
      var _this2 = this;
      debug('authenticate %s', user);
      this.reload(/*#__PURE__*/function () {
        var _ref2 = _asyncToGenerator(/*#__PURE__*/_regenerator().m(function _callee(err) {
          var passwordValid, start, durationMs, _t;
          return _regenerator().w(function (_context) {
            while (1) switch (_context.p = _context.n) {
              case 0:
                debug('reloaded');
                if (!err) {
                  _context.n = 1;
                  break;
                }
                debug('error %o', err);
                return _context.a(2, cb(err.code === 'ENOENT' ? null : err));
              case 1:
                if (_this2.users[user]) {
                  _context.n = 2;
                  break;
                }
                debug('user %s not found', user);
                return _context.a(2, cb(null, false));
              case 2:
                passwordValid = false;
                _context.p = 3;
                start = new Date();
                _context.n = 4;
                return (0, _utils.verifyPassword)(password, _this2.users[user]);
              case 4:
                passwordValid = _context.v;
                durationMs = new Date().getTime() - start.getTime();
                if (durationMs > _this2.slowVerifyMs) {
                  debug('password for user "%s" took %sms to verify', user, durationMs);
                  _this2.logger.warn({
                    user: user,
                    durationMs: durationMs
                  }, 'Password for user "@{user}" took @{durationMs}ms to verify');
                }
                _context.n = 6;
                break;
              case 5:
                _context.p = 5;
                _t = _context.v;
                _this2.logger.error({
                  message: _t.message
                }, 'Unable to verify user password: @{message}');
              case 6:
                if (passwordValid) {
                  _context.n = 7;
                  break;
                }
                debug('password invalid for %s', user);
                return _context.a(2, cb(null, false));
              case 7:
                return _context.a(2, cb(null, [user]));
            }
          }, _callee, null, [[3, 5]]);
        }));
        return function (_x) {
          return _ref2.apply(this, arguments);
        };
      }());
    }

    /**
     * Add user
     * 1. lock file for writing (other processes can still read)
     * 2. reload .htpasswd
     * 3. write new data into .htpasswd.tmp
     * 4. move .htpasswd.tmp to .htpasswd
     * 5. reload .htpasswd
     * 6. unlock file
     *
     * @param {string} user
     * @param {string} password
     * @param {function} realCb
     * @returns {Promise<any>}
     */
  }, {
    key: "adduser",
    value: (function () {
      var _adduser = _asyncToGenerator(/*#__PURE__*/_regenerator().m(function _callee3(user, password, realCb) {
        var _this3 = this;
        var pathPass, sanity;
        return _regenerator().w(function (_context3) {
          while (1) switch (_context3.n) {
            case 0:
              pathPass = this.path;
              debug('adduser %s', user);
              _context3.n = 1;
              return (0, _utils.sanityCheck)(user, password, _utils.verifyPassword, this.users, this.maxUsers);
            case 1:
              sanity = _context3.v;
              debug('sanity check: %s', sanity);
              // preliminary checks, just to ensure that file won't be reloaded if it's
              // not needed
              if (!sanity) {
                _context3.n = 2;
                break;
              }
              debug('sanity check failed');
              return _context3.a(2, realCb(sanity, false));
            case 2:
              (0, _utils.lockAndRead)(pathPass, /*#__PURE__*/function () {
                var _ref3 = _asyncToGenerator(/*#__PURE__*/_regenerator().m(function _callee2(err, res) {
                  var locked, cb, body, _t2, _t3;
                  return _regenerator().w(function (_context2) {
                    while (1) switch (_context2.p = _context2.n) {
                      case 0:
                        locked = false;
                        debug('locked and read');

                        // callback that cleans up lock first
                        cb = function cb(err) {
                          if (locked) {
                            (0, _fileLocking.unlockFile)(pathPass, function () {
                              // ignore any error from the unlock
                              realCb(err, !err);
                            });
                          } else {
                            realCb(err, !err);
                          }
                        };
                        if (!err) {
                          debug('locked');
                          locked = true;
                        }

                        // ignore ENOENT errors, we'll just create .htpasswd in that case
                        if (!(err && err.code !== 'ENOENT')) {
                          _context2.n = 1;
                          break;
                        }
                        return _context2.a(2, cb(err));
                      case 1:
                        debug('read file');
                        body = (res || '').toString('utf8');
                        _this3.users = (0, _utils.parseHTPasswd)(body);
                        debug('parsed users');
                        // real checks, to prevent race conditions
                        // parsing users after reading file.
                        _context2.n = 2;
                        return (0, _utils.sanityCheck)(user, password, _utils.verifyPassword, _this3.users, _this3.maxUsers);
                      case 2:
                        sanity = _context2.v;
                        debug('sanity check: %s', sanity);
                        if (!sanity) {
                          _context2.n = 3;
                          break;
                        }
                        debug('sanity check failed');
                        return _context2.a(2, cb(sanity));
                      case 3:
                        _context2.p = 3;
                        debug('add user to htpasswd file');
                        _t2 = _this3;
                        _context2.n = 4;
                        return (0, _utils.addUserToHTPasswd)(body, user, password, _this3.hashConfig);
                      case 4:
                        _t2._writeFile.call(_t2, _context2.v, cb);
                        debug('user added');
                        _context2.n = 6;
                        break;
                      case 5:
                        _context2.p = 5;
                        _t3 = _context2.v;
                        debug('error %o', _t3);
                        return _context2.a(2, cb(_t3));
                      case 6:
                        return _context2.a(2);
                    }
                  }, _callee2, null, [[3, 5]]);
                }));
                return function (_x5, _x6) {
                  return _ref3.apply(this, arguments);
                };
              }());
            case 3:
              return _context3.a(2);
          }
        }, _callee3, this);
      }));
      function adduser(_x2, _x3, _x4) {
        return _adduser.apply(this, arguments);
      }
      return adduser;
    }()
    /**
     * Reload users
     * @param {function} callback
     */
    )
  }, {
    key: "reload",
    value: function reload(callback) {
      var _this4 = this;
      debug('reload users');
      debug('path: %s', this.path);
      (0, _nodeFs.stat)(this.path, function (err, stats) {
        if (err) {
          return callback(err);
        }
        if (_this4.lastTime === stats.mtime) {
          return callback();
        }
        _this4.lastTime = stats.mtime;
        (0, _nodeFs.readFile)(_this4.path, 'utf8', function (err, buffer) {
          if (err) {
            return callback(err);
          }
          debug('reload users total: %s', Object.keys(_this4.users).length);
          Object.assign(_this4.users, (0, _utils.parseHTPasswd)(buffer));
          callback();
        });
      });
    }
  }, {
    key: "_writeFile",
    value: function _writeFile(body, cb) {
      var _this5 = this;
      (0, _nodeFs.writeFile)(this.path, body, function (err) {
        if (err) {
          cb(err);
        } else {
          _this5.reload(function () {
            cb(null);
          });
        }
      });
    }

    /**
     * changePassword - change password for existing user.
     * @param {string} user
     * @param {string} password
     * @param {string} newPassword
     * @param {function} realCb
     * @returns {function}
     */
  }, {
    key: "changePassword",
    value: function changePassword(user, password, newPassword, realCb) {
      var _this6 = this;
      debug('change password %s', user);
      (0, _utils.lockAndRead)(this.path, /*#__PURE__*/function () {
        var _ref4 = _asyncToGenerator(/*#__PURE__*/_regenerator().m(function _callee4(err, res) {
          var locked, pathPassFile, cb, body, _t4, _t5;
          return _regenerator().w(function (_context4) {
            while (1) switch (_context4.p = _context4.n) {
              case 0:
                debug('locked and read');
                locked = false;
                pathPassFile = _this6.path; // callback that cleans up lock first
                cb = function cb(err) {
                  if (locked) {
                    (0, _fileLocking.unlockFile)(pathPassFile, function () {
                      // ignore any error from the unlock
                      realCb(err, !err);
                    });
                  } else {
                    realCb(err, !err);
                  }
                };
                if (!err) {
                  locked = true;
                }
                if (!(err && err.code !== 'ENOENT')) {
                  _context4.n = 1;
                  break;
                }
                return _context4.a(2, cb(err));
              case 1:
                body = (0, _utils.stringToUtf8)(res);
                _this6.users = (0, _utils.parseHTPasswd)(body);
                _context4.p = 2;
                debug('change password for user %s', user);
                _t4 = _this6;
                _context4.n = 3;
                return (0, _utils.changePasswordToHTPasswd)(body, user, password, newPassword, _this6.hashConfig);
              case 3:
                _t4._writeFile.call(_t4, _context4.v, cb);
                _context4.n = 5;
                break;
              case 4:
                _context4.p = 4;
                _t5 = _context4.v;
                debug('error changing password %o', _t5);
                return _context4.a(2, cb(_t5));
              case 5:
                return _context4.a(2);
            }
          }, _callee4, null, [[2, 4]]);
        }));
        return function (_x7, _x8) {
          return _ref4.apply(this, arguments);
        };
      }());
    }
  }]);
}(_core.pluginUtils.Plugin);
