// verdaccio-htpasswd/build/utils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DEFAULT_BCRYPT_ROUNDS = void 0;
exports.addUserToHTPasswd = addUserToHTPasswd;
exports.changePasswordToHTPasswd = changePasswordToHTPasswd;
exports.generateHtpasswdLine = generateHtpasswdLine;
exports.lockAndRead = lockAndRead;
exports.parseHTPasswd = parseHTPasswd;
exports.sanityCheck = sanityCheck;
exports.stringToUtf8 = stringToUtf8;
exports.verifyPassword = verifyPassword;
var _apacheMd = _interopRequireDefault(require("apache-md5"));
var _bcryptjs = _interopRequireDefault(require("bcryptjs"));
var _debug = _interopRequireDefault(require("debug"));
var _httpErrors = _interopRequireDefault(require("http-errors"));
var _nodeCrypto = _interopRequireDefault(require("node:crypto"));
var _core = require("@verdaccio/core");
var _fileLocking = require("@verdaccio/file-locking");
var _crypt = _interopRequireDefault(require("./crypt3"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { "default": e }; }
function _slicedToArray(r, e) { return _arrayWithHoles(r) || _iterableToArrayLimit(r, e) || _unsupportedIterableToArray(r, e) || _nonIterableRest(); }
function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _unsupportedIterableToArray(r, a) { if (r) { if ("string" == typeof r) return _arrayLikeToArray(r, a); var t = {}.toString.call(r).slice(8, -1); return "Object" === t && r.constructor && (t = r.constructor.name), "Map" === t || "Set" === t ? Array.from(r) : "Arguments" === t || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(t) ? _arrayLikeToArray(r, a) : void 0; } }
function _arrayLikeToArray(r, a) { (null == a || a > r.length) && (a = r.length); for (var e = 0, n = Array(a); e < a; e++) n[e] = r[e]; return n; }
function _iterableToArrayLimit(r, l) { var t = null == r ? null : "undefined" != typeof Symbol && r[Symbol.iterator] || r["@@iterator"]; if (null != t) { var e, n, i, u, a = [], f = !0, o = !1; try { if (i = (t = t.call(r)).next, 0 === l) { if (Object(t) !== t) return; f = !1; } else for (; !(f = (e = i.call(t)).done) && (a.push(e.value), a.length !== l); f = !0); } catch (r) { o = !0, n = r; } finally { try { if (!f && null != t["return"] && (u = t["return"](), Object(u) !== u)) return; } finally { if (o) throw n; } } return a; } }
function _arrayWithHoles(r) { if (Array.isArray(r)) return r; }
function _regenerator() { /*! regenerator-runtime -- Copyright (c) 2014-present, Facebook, Inc. -- license (MIT): https://github.com/babel/babel/blob/main/packages/babel-helpers/LICENSE */ var e, t, r = "function" == typeof Symbol ? Symbol : {}, n = r.iterator || "@@iterator", o = r.toStringTag || "@@toStringTag"; function i(r, n, o, i) { var c = n && n.prototype instanceof Generator ? n : Generator, u = Object.create(c.prototype); return _regeneratorDefine2(u, "_invoke", function (r, n, o) { var i, c, u, f = 0, p = o || [], y = !1, G = { p: 0, n: 0, v: e, a: d, f: d.bind(e, 4), d: function d(t, r) { return i = t, c = 0, u = e, G.n = r, a; } }; function d(r, n) { for (c = r, u = n, t = 0; !y && f && !o && t < p.length; t++) { var o, i = p[t], d = G.p, l = i[2]; r > 3 ? (o = l === n) && (u = i[(c = i[4]) ? 5 : (c = 3, 3)], i[4] = i[5] = e) : i[0] <= d && ((o = r < 2 && d < i[1]) ? (c = 0, G.v = n, G.n = i[1]) : d < l && (o = r < 3 || i[0] > n || n > l) && (i[4] = r, i[5] = n, G.n = l, c = 0)); } if (o || r > 1) return a; throw y = !0, n; } return function (o, p, l) { if (f > 1) throw TypeError("Generator is already running"); for (y && 1 === p && d(p, l), c = p, u = l; (t = c < 2 ? e : u) || !y;) { i || (c ? c < 3 ? (c > 1 && (G.n = -1), d(c, u)) : G.n = u : G.v = u); try { if (f = 2, i) { if (c || (o = "next"), t = i[o]) { if (!(t = t.call(i, u))) throw TypeError("iterator result is not an object"); if (!t.done) return t; u = t.value, c < 2 && (c = 0); } else 1 === c && (t = i["return"]) && t.call(i), c < 2 && (u = TypeError("The iterator does not provide a '" + o + "' method"), c = 1); i = e; } else if ((t = (y = G.n < 0) ? u : r.call(n, G)) !== a) break; } catch (t) { i = e, c = 1, u = t; } finally { f = 1; } } return { value: t, done: y }; }; }(r, o, i), !0), u; } var a = {}; function Generator() {} function GeneratorFunction() {} function GeneratorFunctionPrototype() {} t = Object.getPrototypeOf; var c = [][n] ? t(t([][n]())) : (_regeneratorDefine2(t = {}, n, function () { return this; }), t), u = GeneratorFunctionPrototype.prototype = Generator.prototype = Object.create(c); function f(e) { return Object.setPrototypeOf ? Object.setPrototypeOf(e, GeneratorFunctionPrototype) : (e.__proto__ = GeneratorFunctionPrototype, _regeneratorDefine2(e, o, "GeneratorFunction")), e.prototype = Object.create(u), e; } return GeneratorFunction.prototype = GeneratorFunctionPrototype, _regeneratorDefine2(u, "constructor", GeneratorFunctionPrototype), _regeneratorDefine2(GeneratorFunctionPrototype, "constructor", GeneratorFunction), GeneratorFunction.displayName = "GeneratorFunction", _regeneratorDefine2(GeneratorFunctionPrototype, o, "GeneratorFunction"), _regeneratorDefine2(u), _regeneratorDefine2(u, o, "Generator"), _regeneratorDefine2(u, n, function () { return this; }), _regeneratorDefine2(u, "toString", function () { return "[object Generator]"; }), (_regenerator = function _regenerator() { return { w: i, m: f }; })(); }
function _regeneratorDefine2(e, r, n, t) { var i = Object.defineProperty; try { i({}, "", {}); } catch (e) { i = 0; } _regeneratorDefine2 = function _regeneratorDefine(e, r, n, t) { function o(r, n) { _regeneratorDefine2(e, r, function (e) { return this._invoke(r, n, e); }); } r ? i ? i(e, r, { value: n, enumerable: !t, configurable: !t, writable: !t }) : e[r] = n : (o("next", 0), o("throw", 1), o("return", 2)); }, _regeneratorDefine2(e, r, n, t); }
function asyncGeneratorStep(n, t, e, r, o, a, c) { try { var i = n[a](c), u = i.value; } catch (n) { return void e(n); } i.done ? t(u) : Promise.resolve(u).then(r, o); }
function _asyncToGenerator(n) { return function () { var t = this, e = arguments; return new Promise(function (r, o) { var a = n.apply(t, e); function _next(n) { asyncGeneratorStep(a, r, o, _next, _throw, "next", n); } function _throw(n) { asyncGeneratorStep(a, r, o, _next, _throw, "throw", n); } _next(void 0); }); }; }
var debug = (0, _debug["default"])('verdaccio:plugin:htpasswd:utils');
var DEFAULT_BCRYPT_ROUNDS = exports.DEFAULT_BCRYPT_ROUNDS = 10;
// this function neither unlocks file nor closes it
// it'll have to be done manually later
function lockAndRead(name, cb) {
  (0, _fileLocking.readFile)(name, {
    lock: true
  }, function (err, res) {
    if (err) {
      return cb(err);
    }
    return cb(null, res);
  });
}

/**
 * parseHTPasswd - convert htpasswd lines to object.
 * @param {string} input
 * @returns {object}
 */
function parseHTPasswd(input) {
  // The input is split on line ending styles that are both windows and unix compatible
  return input.split(/[\r]?[\n]/).reduce(function (result, line) {
    var args = line.split(':', 3).map(function (str) {
      return str.trim();
    });
    if (args.length > 1) {
      result[args[0]] = args[1];
    }
    return result;
  }, {});
}

/**
 * verifyPassword - matches password and it's hash.
 * @param {string} passwd
 * @param {string} hash
 * @returns {Promise<boolean>}
 */
function verifyPassword(_x, _x2) {
  return _verifyPassword.apply(this, arguments);
}
/**
 * generateHtpasswdLine - generates line for htpasswd file.
 * @param {string} user
 * @param {string} passwd
 * @param {HtpasswdHashConfig} hashConfig
 * @returns {string}
 */
function _verifyPassword() {
  _verifyPassword = _asyncToGenerator(/*#__PURE__*/_regenerator().m(function _callee(passwd, hash) {
    return _regenerator().w(function (_context) {
      while (1) switch (_context.n) {
        case 0:
          if (!hash.match(/^\$2([aby])\$/)) {
            _context.n = 2;
            break;
          }
          _context.n = 1;
          return _bcryptjs["default"].compare(passwd, hash);
        case 1:
          return _context.a(2, _context.v);
        case 2:
          if (!(hash.indexOf('{PLAIN}') === 0)) {
            _context.n = 3;
            break;
          }
          return _context.a(2, passwd === hash.slice(7));
        case 3:
          if (!(hash.indexOf('{SHA}') === 0)) {
            _context.n = 4;
            break;
          }
          return _context.a(2, _nodeCrypto["default"].createHash('sha1')
          // https://nodejs.org/api/crypto.html#crypto_hash_update_data_inputencoding
          .update(passwd, 'utf8').digest('base64') === hash.slice(5));
        case 4:
          return _context.a(2, (0, _apacheMd["default"])(passwd, hash) === hash || (0, _crypt["default"])(passwd, hash) === hash);
      }
    }, _callee);
  }));
  return _verifyPassword.apply(this, arguments);
}
function generateHtpasswdLine(_x3, _x4, _x5) {
  return _generateHtpasswdLine.apply(this, arguments);
}
/**
 * addUserToHTPasswd - Generate a htpasswd format for .htpasswd
 * @param {string} body
 * @param {string} user
 * @param {string} passwd
 * @param {HtpasswdHashConfig} hashConfig
 * @returns {string}
 */
function _generateHtpasswdLine() {
  _generateHtpasswdLine = _asyncToGenerator(/*#__PURE__*/_regenerator().m(function _callee2(user, passwd, hashConfig) {
    var hash, comment, _t;
    return _regenerator().w(function (_context2) {
      while (1) switch (_context2.n) {
        case 0:
          debug('algorithm %o', hashConfig.algorithm);
          _t = hashConfig.algorithm;
          _context2.n = _t === _core.constants.HtpasswdHashAlgorithm.bcrypt ? 1 : _t === _core.constants.HtpasswdHashAlgorithm.crypt ? 3 : _t === _core.constants.HtpasswdHashAlgorithm.md5 ? 4 : _t === _core.constants.HtpasswdHashAlgorithm.sha1 ? 5 : 6;
          break;
        case 1:
          _context2.n = 2;
          return _bcryptjs["default"].hash(passwd, hashConfig.rounds || DEFAULT_BCRYPT_ROUNDS);
        case 2:
          hash = _context2.v;
          return _context2.a(3, 7);
        case 3:
          hash = (0, _crypt["default"])(passwd);
          return _context2.a(3, 7);
        case 4:
          hash = (0, _apacheMd["default"])(passwd);
          return _context2.a(3, 7);
        case 5:
          hash = '{SHA}' + _nodeCrypto["default"].createHash('sha1').update(passwd, 'utf8').digest('base64');
          return _context2.a(3, 7);
        case 6:
          throw (0, _httpErrors["default"])('Unexpected hash algorithm');
        case 7:
          comment = 'autocreated ' + new Date().toJSON();
          return _context2.a(2, "".concat(user, ":").concat(hash, ":").concat(comment, "\n"));
      }
    }, _callee2);
  }));
  return _generateHtpasswdLine.apply(this, arguments);
}
function addUserToHTPasswd(_x6, _x7, _x8, _x9) {
  return _addUserToHTPasswd.apply(this, arguments);
}
/**
 * Sanity check for a user
 * @param {string} user
 * @param {object} users
 * @param {string} password
 * @param {Callback} verifyFn
 * @param {number} maxUsers
 * @returns {object}
 */
function _addUserToHTPasswd() {
  _addUserToHTPasswd = _asyncToGenerator(/*#__PURE__*/_regenerator().m(function _callee3(body, user, passwd, hashConfig) {
    var err, newline;
    return _regenerator().w(function (_context3) {
      while (1) switch (_context3.n) {
        case 0:
          if (!(user !== encodeURIComponent(user))) {
            _context3.n = 1;
            break;
          }
          err = (0, _httpErrors["default"])('username should not contain non-uri-safe characters');
          err.status = _core.HTTP_STATUS.CONFLICT;
          throw err;
        case 1:
          _context3.n = 2;
          return generateHtpasswdLine(user, passwd, hashConfig);
        case 2:
          newline = _context3.v;
          if (body.length && body[body.length - 1] !== '\n') {
            newline = '\n' + newline;
          }
          return _context3.a(2, body + newline);
      }
    }, _callee3);
  }));
  return _addUserToHTPasswd.apply(this, arguments);
}
function sanityCheck(_x0, _x1, _x10, _x11, _x12) {
  return _sanityCheck.apply(this, arguments);
}
/**
 * /**
 * changePasswordToHTPasswd - change password for existing user
 * @param {string} body
 * @param {string} user
 * @param {string} passwd
 * @param {string} newPasswd
 * @param {HtpasswdHashConfig} hashConfig
 * @returns {Promise<string>}
 */
function _sanityCheck() {
  _sanityCheck = _asyncToGenerator(/*#__PURE__*/_regenerator().m(function _callee4(user, password, verifyFn, users, maxUsers) {
    var err, hash, auth;
    return _regenerator().w(function (_context4) {
      while (1) switch (_context4.n) {
        case 0:
          if (!(!user || !password)) {
            _context4.n = 1;
            break;
          }
          debug('username or password is missing');
          err = Error(_core.API_ERROR.USERNAME_PASSWORD_REQUIRED);
          err.status = _core.HTTP_STATUS.BAD_REQUEST;
          return _context4.a(2, err);
        case 1:
          hash = users[user];
          if (!(maxUsers < 0)) {
            _context4.n = 2;
            break;
          }
          debug('registration is disabled');
          err = Error(_core.API_ERROR.REGISTRATION_DISABLED);
          err.status = _core.HTTP_STATUS.CONFLICT;
          return _context4.a(2, err);
        case 2:
          if (!hash) {
            _context4.n = 5;
            break;
          }
          _context4.n = 3;
          return verifyFn(password, users[user]);
        case 3:
          auth = _context4.v;
          if (!auth) {
            _context4.n = 4;
            break;
          }
          debug("user ".concat(user, " already exists"));
          err = Error(_core.API_ERROR.USERNAME_ALREADY_REGISTERED);
          err.status = _core.HTTP_STATUS.CONFLICT;
          return _context4.a(2, err);
        case 4:
          debug("user ".concat(user, " exists but password is wrong"));
          err = Error(_core.API_ERROR.UNAUTHORIZED_ACCESS);
          err.status = _core.HTTP_STATUS.UNAUTHORIZED;
          return _context4.a(2, err);
        case 5:
          if (!(Object.keys(users).length >= maxUsers)) {
            _context4.n = 6;
            break;
          }
          debug('maximum amount of users reached');
          err = Error(_core.API_ERROR.MAX_USERS_REACHED);
          err.status = _core.HTTP_STATUS.FORBIDDEN;
          return _context4.a(2, err);
        case 6:
          debug('sanity check passed');
          return _context4.a(2, null);
      }
    }, _callee4);
  }));
  return _sanityCheck.apply(this, arguments);
}
function changePasswordToHTPasswd(_x13, _x14, _x15, _x16, _x17) {
  return _changePasswordToHTPasswd.apply(this, arguments);
}
function _changePasswordToHTPasswd() {
  _changePasswordToHTPasswd = _asyncToGenerator(/*#__PURE__*/_regenerator().m(function _callee5(body, user, passwd, newPasswd, hashConfig) {
    var lines, userLineIndex, _lines$userLineIndex$, _lines$userLineIndex$2, username, hash, passwordValid, updatedUserLine;
    return _regenerator().w(function (_context5) {
      while (1) switch (_context5.n) {
        case 0:
          debug('change password for user %o', user);
          lines = body.split('\n');
          userLineIndex = lines.findIndex(function (line) {
            return line.split(':', 1).shift() === user;
          });
          if (!(userLineIndex === -1)) {
            _context5.n = 1;
            break;
          }
          debug('user %o does not exist', user);
          throw new Error("Unable to change password for user '".concat(user, "': user does not currently exist"));
        case 1:
          _lines$userLineIndex$ = lines[userLineIndex].split(':', 2), _lines$userLineIndex$2 = _slicedToArray(_lines$userLineIndex$, 2), username = _lines$userLineIndex$2[0], hash = _lines$userLineIndex$2[1];
          _context5.n = 2;
          return verifyPassword(passwd, hash);
        case 2:
          passwordValid = _context5.v;
          if (passwordValid) {
            _context5.n = 3;
            break;
          }
          debug("invalid old password");
          throw new Error("Unable to change password for user '".concat(user, "': invalid old password"));
        case 3:
          _context5.n = 4;
          return generateHtpasswdLine(username, newPasswd, hashConfig);
        case 4:
          updatedUserLine = _context5.v;
          lines.splice(userLineIndex, 1, updatedUserLine);
          debug('password changed');
          return _context5.a(2, lines.join('\n'));
      }
    }, _callee5);
  }));
  return _changePasswordToHTPasswd.apply(this, arguments);
}
function stringToUtf8(authentication) {
  return (authentication || '').toString();
}
