// @verdaccio/local-storage-legacy/lib/local-database.js
const require_runtime = require("./_virtual/_rolldown/runtime.js");
const require_local_fs = require("./local-fs.js");
const require_pkg_utils = require("./pkg-utils.js");
const require_token = require("./token.js");
let debug = require("debug");
debug = require_runtime.__toESM(debug);
let globby = require("globby");
globby = require_runtime.__toESM(globby);
let lodash = require("lodash");
lodash = require_runtime.__toESM(lodash);
let mkdirp = require("mkdirp");
mkdirp = require_runtime.__toESM(mkdirp);
let node_fs = require("node:fs");
node_fs = require_runtime.__toESM(node_fs);
let node_path = require("node:path");
node_path = require_runtime.__toESM(node_path);
let sanitize_filename = require("sanitize-filename");
sanitize_filename = require_runtime.__toESM(sanitize_filename);
let _verdaccio_core = require("@verdaccio/core");
//#region src/local-database.ts
var DEPRECATED_DB_NAME = ".sinopia-db.json";
var DB_NAME = ".verdaccio-db.json";
var debug$1 = (0, debug.default)("verdaccio:plugin:local-storage-legacy");
var LocalDatabase = class extends require_token.default {
	path;
	logger;
	data;
	config;
	locked;
	constructor(config, logger) {
		super(config);
		this.config = config;
		this.path = this._buildStoragePath(config);
		this.logger = logger;
		this.locked = false;
		this.data = this._fetchLocalPackages();
		this._sync();
	}
	getSecret() {
		return Promise.resolve(this.data.secret);
	}
	setSecret(secret) {
		return new Promise((resolve) => {
			this.data.secret = secret;
			resolve(this._sync());
		});
	}
	add(name, cb) {
		if (this.data.list.indexOf(name) === -1) {
			this.data.list.push(name);
			debug$1("the private package %o has been added", name);
			cb(this._sync());
		} else {
			debug$1("the private package %o was not added", name);
			cb(null);
		}
	}
	search(onPackage, onEnd, validateName) {
		const storages = this._getCustomPackageLocalStorages();
		debug$1(`search custom local packages: %o`, JSON.stringify(storages));
		const base = node_path.default.dirname(this.config.self_path);
		const self = this;
		const storageKeys = Object.keys(storages);
		debug$1(`search base: %o keys: %o`, base, storageKeys);
		const emitPackage = (item) => new Promise((resolve, reject) => {
			onPackage(item, (err) => err ? reject(err) : resolve());
		});
		const statAsync = (filePath) => new Promise((resolve, reject) => {
			node_fs.default.stat(filePath, (err, stats) => err ? reject(err) : resolve(stats));
		});
		const processSearch = async () => {
			for (const storage of storageKeys) {
				const position = storageKeys.indexOf(storage);
				const base2 = node_path.default.join(position !== 0 ? storageKeys[0] : "");
				const storagePath = node_path.default.resolve(base, base2, storage);
				debug$1("search path: %o : %o", storagePath, storage);
				const folders = await (0, globby.default)("*", {
					cwd: storagePath,
					onlyDirectories: true,
					onlyFiles: false,
					deep: 1,
					dot: false,
					followSymbolicLinks: true
				});
				for (const folder of folders) {
					if (storageKeys.includes(folder)) continue;
					if (folder.match(/^@/)) {
						const scopePath = node_path.default.resolve(storagePath, folder);
						const scopedFolders = await (0, globby.default)("*", {
							cwd: scopePath,
							onlyDirectories: true,
							onlyFiles: false,
							deep: 1,
							dot: false
						});
						for (const scopedPkg of scopedFolders) if (validateName(scopedPkg)) {
							const packagePath = node_path.default.resolve(scopePath, scopedPkg);
							const stats = await statAsync(packagePath);
							await emitPackage({
								name: `${folder}/${scopedPkg}`,
								path: packagePath,
								time: stats.mtime.getTime()
							});
						}
					} else if (validateName(folder)) {
						const packagePath = node_path.default.resolve(storagePath, folder);
						debug$1("search file location: %o", packagePath);
						const stats = await statAsync(packagePath);
						await emitPackage({
							name: folder,
							path: packagePath,
							time: self.getTime(stats.mtime.getTime(), stats.mtime)
						});
					}
				}
			}
		};
		processSearch().then(() => onEnd(null)).catch((err) => onEnd(err));
	}
	remove(name, cb) {
		this.get((err, data) => {
			if (err) {
				cb(_verdaccio_core.errorUtils.getInternalError("error remove private package"));
				this.logger.error({ err }, "remove the private package has failed @{err}");
				debug$1("error on remove package %o", name);
			}
			const pkgName = data.indexOf(name);
			if (pkgName !== -1) {
				this.data.list.splice(pkgName, 1);
				debug$1("remove package %o has been removed", name);
			}
			cb(this._sync());
		});
	}
	/**
	* Return all database elements.
	* @return {Array}
	*/
	get(cb) {
		const list = this.data.list;
		const totalItems = this.data.list.length;
		cb(null, list);
		debug$1("get full list of packages (%o) has been fetched", totalItems);
	}
	getPackageStorage(packageName) {
		const packageAccess = this.config.getMatchedPackagesSpec(packageName);
		const packagePath = this._getLocalStoragePath(packageAccess ? packageAccess.storage : void 0);
		debug$1("storage path selected: ", packagePath);
		if (lodash.default.isString(packagePath) === false) {
			debug$1("the package %o has no storage defined ", packageName);
			return;
		}
		const sanitizedName = packageName.split("/").map((segment) => (0, sanitize_filename.default)(segment)).join("/");
		const packageStoragePath = node_path.default.join(node_path.default.resolve(node_path.default.dirname(this.config.self_path || ""), packagePath), sanitizedName);
		debug$1("storage absolute path: ", packageStoragePath);
		return new require_local_fs.default(packageStoragePath, this.logger);
	}
	clean() {
		this._sync();
	}
	getTime(time, mtime) {
		return time ? time : mtime;
	}
	_getCustomPackageLocalStorages() {
		const storages = {};
		if (this.config.storage) storages[this.config.storage] = true;
		const { packages } = this.config;
		if (packages) Object.keys(packages || {}).map((pkg) => {
			const storage = packages[pkg].storage;
			if (storage) storages[storage] = false;
		});
		return storages;
	}
	/**
	* Syncronize {create} database whether does not exist.
	* @return {Error|*}
	*/
	_sync() {
		debug$1("sync database started");
		if (this.locked) {
			this.logger.error("Database is locked, please check error message printed during startup to prevent data loss.");
			return /* @__PURE__ */ new Error("Verdaccio database is locked, please contact your administrator to checkout logs during verdaccio startup.");
		}
		try {
			const folderName = node_path.default.dirname(this.path);
			mkdirp.default.sync(folderName);
			debug$1("sync folder %o created succeed", folderName);
		} catch (err) {
			debug$1("sync create folder has failed with error: %o", err);
			return null;
		}
		try {
			node_fs.default.writeFileSync(this.path, JSON.stringify(this.data));
			debug$1("sync write succeed");
			return null;
		} catch (err) {
			debug$1("sync failed %o", err);
			return err;
		}
	}
	/**
	* Verify the right local storage location.
	* @param {String} path
	* @return {String}
	* @private
	*/
	_getLocalStoragePath(storage) {
		const globalConfigStorage = this.config ? this.config.storage : void 0;
		if (lodash.default.isNil(globalConfigStorage)) throw new Error("global storage is required for this plugin");
		else {
			if (lodash.default.isNil(storage) === false && lodash.default.isString(storage)) return node_path.default.join(globalConfigStorage, storage);
			return globalConfigStorage;
		}
	}
	/**
	* Build the local database path.
	* @param {Object} config
	* @return {string|String|*}
	* @private
	*/
	_buildStoragePath(config) {
		const sinopiadbPath = this._dbGenPath(DEPRECATED_DB_NAME, config);
		try {
			node_fs.default.accessSync(sinopiadbPath, node_fs.default.constants.F_OK);
			process.emitWarning("Database name deprecated!", {
				code: "VERCODE01",
				detail: `Please rename database name from ${DEPRECATED_DB_NAME} to ${DB_NAME}`
			});
			return sinopiadbPath;
		} catch (err) {
			if (err.code === "ENOENT") return this._dbGenPath(DB_NAME, config);
			throw err;
		}
	}
	/**
	* Fetch local packages.
	* @private
	* @return {Object}
	*/
	_fetchLocalPackages() {
		const emptyDatabase = {
			list: [],
			secret: ""
		};
		try {
			return require_pkg_utils.loadPrivatePackages(this.path, this.logger);
		} catch (err) {
			if (err.code !== "ENOENT") {
				this.locked = true;
				this.logger.error("Failed to read package database file, please check the error printed below:\n", `File Path: ${this.path}\n\n ${err.message}`);
			}
			return emptyDatabase;
		}
	}
};
//#endregion
exports.default = LocalDatabase;
