// @verdaccio/loaders/build/plugin-async-loader.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.asyncLoadPlugin = asyncLoadPlugin;
exports.executePlugin = executePlugin;
var _debug = _interopRequireDefault(require("debug"));
var _lodash = _interopRequireDefault(require("lodash"));
var _promises = require("node:fs/promises");
var _nodePath = require("node:path");
var _core = require("@verdaccio/core");
var _utils = require("./utils");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { "default": e }; }
function _regenerator() { /*! regenerator-runtime -- Copyright (c) 2014-present, Facebook, Inc. -- license (MIT): https://github.com/babel/babel/blob/main/packages/babel-helpers/LICENSE */ var e, t, r = "function" == typeof Symbol ? Symbol : {}, n = r.iterator || "@@iterator", o = r.toStringTag || "@@toStringTag"; function i(r, n, o, i) { var c = n && n.prototype instanceof Generator ? n : Generator, u = Object.create(c.prototype); return _regeneratorDefine2(u, "_invoke", function (r, n, o) { var i, c, u, f = 0, p = o || [], y = !1, G = { p: 0, n: 0, v: e, a: d, f: d.bind(e, 4), d: function d(t, r) { return i = t, c = 0, u = e, G.n = r, a; } }; function d(r, n) { for (c = r, u = n, t = 0; !y && f && !o && t < p.length; t++) { var o, i = p[t], d = G.p, l = i[2]; r > 3 ? (o = l === n) && (u = i[(c = i[4]) ? 5 : (c = 3, 3)], i[4] = i[5] = e) : i[0] <= d && ((o = r < 2 && d < i[1]) ? (c = 0, G.v = n, G.n = i[1]) : d < l && (o = r < 3 || i[0] > n || n > l) && (i[4] = r, i[5] = n, G.n = l, c = 0)); } if (o || r > 1) return a; throw y = !0, n; } return function (o, p, l) { if (f > 1) throw TypeError("Generator is already running"); for (y && 1 === p && d(p, l), c = p, u = l; (t = c < 2 ? e : u) || !y;) { i || (c ? c < 3 ? (c > 1 && (G.n = -1), d(c, u)) : G.n = u : G.v = u); try { if (f = 2, i) { if (c || (o = "next"), t = i[o]) { if (!(t = t.call(i, u))) throw TypeError("iterator result is not an object"); if (!t.done) return t; u = t.value, c < 2 && (c = 0); } else 1 === c && (t = i["return"]) && t.call(i), c < 2 && (u = TypeError("The iterator does not provide a '" + o + "' method"), c = 1); i = e; } else if ((t = (y = G.n < 0) ? u : r.call(n, G)) !== a) break; } catch (t) { i = e, c = 1, u = t; } finally { f = 1; } } return { value: t, done: y }; }; }(r, o, i), !0), u; } var a = {}; function Generator() {} function GeneratorFunction() {} function GeneratorFunctionPrototype() {} t = Object.getPrototypeOf; var c = [][n] ? t(t([][n]())) : (_regeneratorDefine2(t = {}, n, function () { return this; }), t), u = GeneratorFunctionPrototype.prototype = Generator.prototype = Object.create(c); function f(e) { return Object.setPrototypeOf ? Object.setPrototypeOf(e, GeneratorFunctionPrototype) : (e.__proto__ = GeneratorFunctionPrototype, _regeneratorDefine2(e, o, "GeneratorFunction")), e.prototype = Object.create(u), e; } return GeneratorFunction.prototype = GeneratorFunctionPrototype, _regeneratorDefine2(u, "constructor", GeneratorFunctionPrototype), _regeneratorDefine2(GeneratorFunctionPrototype, "constructor", GeneratorFunction), GeneratorFunction.displayName = "GeneratorFunction", _regeneratorDefine2(GeneratorFunctionPrototype, o, "GeneratorFunction"), _regeneratorDefine2(u), _regeneratorDefine2(u, o, "Generator"), _regeneratorDefine2(u, n, function () { return this; }), _regeneratorDefine2(u, "toString", function () { return "[object Generator]"; }), (_regenerator = function _regenerator() { return { w: i, m: f }; })(); }
function _regeneratorDefine2(e, r, n, t) { var i = Object.defineProperty; try { i({}, "", {}); } catch (e) { i = 0; } _regeneratorDefine2 = function _regeneratorDefine(e, r, n, t) { function o(r, n) { _regeneratorDefine2(e, r, function (e) { return this._invoke(r, n, e); }); } r ? i ? i(e, r, { value: n, enumerable: !t, configurable: !t, writable: !t }) : e[r] = n : (o("next", 0), o("throw", 1), o("return", 2)); }, _regeneratorDefine2(e, r, n, t); }
function asyncGeneratorStep(n, t, e, r, o, a, c) { try { var i = n[a](c), u = i.value; } catch (n) { return void e(n); } i.done ? t(u) : Promise.resolve(u).then(r, o); }
function _asyncToGenerator(n) { return function () { var t = this, e = arguments; return new Promise(function (r, o) { var a = n.apply(t, e); function _next(n) { asyncGeneratorStep(a, r, o, _next, _throw, "next", n); } function _throw(n) { asyncGeneratorStep(a, r, o, _next, _throw, "throw", n); } _next(void 0); }); }; }
var debug = (0, _debug["default"])('verdaccio:plugin:loader:async');
function isDirectory(_x) {
  return _isDirectory.apply(this, arguments);
}
function _isDirectory() {
  _isDirectory = _asyncToGenerator(/*#__PURE__*/_regenerator().m(function _callee(pathFolder) {
    var stat;
    return _regenerator().w(function (_context) {
      while (1) switch (_context.n) {
        case 0:
          _context.n = 1;
          return (0, _promises.lstat)(pathFolder);
        case 1:
          stat = _context.v;
          return _context.a(2, stat.isDirectory());
      }
    }, _callee);
  }));
  return _isDirectory.apply(this, arguments);
}
function mergeConfig(appConfig, pluginConfig) {
  return _lodash["default"].merge({}, appConfig, pluginConfig);
}

// type Plugins<T> =
//   | pluginUtils.Auth<T>
//   | pluginUtils.Storage<T>
//   | pluginUtils.ExpressMiddleware<T, unknown, unknown>;

/**
 * The plugin loader find recursively plugins, if one plugin fails is ignored and report the error to the logger.
 *
 * The loader follows the order:
 * - If the at the `config.yaml` file the  `plugins: ./plugins` is defined
 *   - If is absolute will use the provided path
 *   - If is relative, will use the base path of the config file. eg: /root/config.yaml the plugins folder should be
 *     hosted at /root/plugins
 * - The next step is find at the node_modules or global based on the `require` native algorithm.
 * - If the package is scoped eg: @scope/foo, try to load the package `@scope/foo`
 * - If the package is not scoped, will use the default prefix: verdaccio-foo ("verdaccio-theme-" prefix for theme ui plugins).
 * - If a custom prefix is provided, the verdaccio- is replaced by the config.server.pluginPrefix.
 *
 * The `sanityCheck` is the validation for the required methods to load the plugin, if the validation fails the plugin won't be loaded.
 * The `params` is an object that contains the global configuration and the logger.
 *
 * @param {*} pluginConfigs the custom plugin section
 * @param {*} pluginOptions a set of options to initialize the plugin
 * @param {*} sanityCheck callback that check the shape that should fulfill the plugin
 * @param {*} prefix by default is verdaccio but can be override with config.server.pluginPrefix
 * @param {*} pluginCategory the category of the plugin, eg: auth, storage, middleware
 * @return {Array} list of plugins
 */
function asyncLoadPlugin() {
  return _asyncLoadPlugin.apply(this, arguments);
}
function _asyncLoadPlugin() {
  _asyncLoadPlugin = _asyncToGenerator(/*#__PURE__*/_regenerator().m(function _callee2() {
    var pluginConfigs,
      pluginOptions,
      sanityCheck,
      legacyMergeConfigs,
      prefix,
      pluginCategory,
      logger,
      pluginsIds,
      config,
      plugins,
      _i,
      _pluginsIds,
      pluginId,
      isScoped,
      pluginName,
      pluginsPath,
      pluginDir,
      externalFilePlugin,
      plugin,
      _plugin,
      _args2 = arguments,
      _t;
    return _regenerator().w(function (_context2) {
      while (1) switch (_context2.p = _context2.n) {
        case 0:
          pluginConfigs = _args2.length > 0 && _args2[0] !== undefined ? _args2[0] : {};
          pluginOptions = _args2.length > 1 ? _args2[1] : undefined;
          sanityCheck = _args2.length > 2 ? _args2[2] : undefined;
          legacyMergeConfigs = _args2.length > 3 && _args2[3] !== undefined ? _args2[3] : false;
          prefix = _args2.length > 4 && _args2[4] !== undefined ? _args2[4] : _core.PLUGIN_PREFIX;
          pluginCategory = _args2.length > 5 && _args2[5] !== undefined ? _args2[5] : 'unknown';
          logger = pluginOptions === null || pluginOptions === void 0 ? void 0 : pluginOptions.logger;
          pluginsIds = Object.keys(pluginConfigs || {});
          config = pluginOptions.config;
          plugins = [];
          _i = 0, _pluginsIds = pluginsIds;
        case 1:
          if (!(_i < _pluginsIds.length)) {
            _context2.n = 13;
            break;
          }
          pluginId = _pluginsIds[_i];
          debug('>>> looking for plugin %o', pluginId);
          isScoped = pluginId.startsWith('@') && pluginId.includes('/');
          debug('is scoped plugin: %s', isScoped);
          pluginName = isScoped ? pluginId : "".concat(prefix, "-").concat(pluginId);
          debug('plugin package name %s', pluginName);

          // Try to load the plugin from the config.plugins path
          if (!(typeof config.plugins === 'string')) {
            _context2.n = 9;
            break;
          }
          pluginsPath = config.plugins;
          debug('plugin path %s', pluginsPath);
          if ((0, _nodePath.isAbsolute)(pluginsPath)) {
            _context2.n = 3;
            break;
          }
          if (typeof config.config_path === 'string' && !config.configPath) {
            logger.error('configPath is missing and the legacy config.config_path is not available for loading plugins');
          }
          if (config.configPath) {
            _context2.n = 2;
            break;
          }
          logger.error('config path property is required for loading plugins');
          return _context2.a(3, 12);
        case 2:
          pluginsPath = (0, _nodePath.resolve)((0, _nodePath.join)((0, _nodePath.dirname)(config.configPath), pluginsPath));
        case 3:
          logger.debug({
            path: pluginsPath
          }, 'plugins folder defined, loading plugins from @{path} ');
          // throws if is not a directory
          _context2.p = 4;
          _context2.n = 5;
          return isDirectory(pluginsPath);
        case 5:
          pluginDir = pluginsPath;
          externalFilePlugin = (0, _nodePath.resolve)(pluginDir, pluginName);
          plugin = (0, _utils.tryLoad)(externalFilePlugin, function (a, b) {
            logger.error(a, b);
          });
          debug('external plugin %o', plugin);
          if (!(plugin && (0, _utils.isValid)(plugin))) {
            _context2.n = 7;
            break;
          }
          plugin = executePlugin(plugin, pluginConfigs[pluginId], pluginOptions, legacyMergeConfigs);
          if (sanityCheck(plugin)) {
            _context2.n = 6;
            break;
          }
          logger.error({
            content: externalFilePlugin
          }, "@{content} doesn't look like a valid plugin");
          return _context2.a(3, 12);
        case 6:
          debug('>>> plugin is running and passed sanity check');
          plugins.push(plugin);
          logger.info({
            pluginName: pluginName,
            pluginCategory: pluginCategory
          }, 'plugin @{pluginName} successfully loaded (@{pluginCategory})');
          return _context2.a(3, 12);
        case 7:
          _context2.n = 9;
          break;
        case 8:
          _context2.p = 8;
          _t = _context2.v;
          logger.warn({
            err: _t.message,
            pluginsPath: pluginsPath,
            pluginName: pluginName
          }, '@{err} on loading plugins at @{pluginsPath} for @{pluginName}');
        case 9:
          if (!(typeof pluginId === 'string')) {
            _context2.n = 12;
            break;
          }
          _plugin = (0, _utils.tryLoad)(pluginName, function (a, b) {
            logger.error(a, b);
          });
          if (!(_plugin && (0, _utils.isValid)(_plugin))) {
            _context2.n = 11;
            break;
          }
          _plugin = executePlugin(_plugin, pluginConfigs[pluginId], pluginOptions, legacyMergeConfigs);
          if (sanityCheck(_plugin)) {
            _context2.n = 10;
            break;
          }
          logger.error({
            pluginName: pluginName
          }, "@{pluginName} doesn't look like a valid plugin");
          return _context2.a(3, 12);
        case 10:
          debug('>>> plugin is running and passed sanity check');
          plugins.push(_plugin);
          logger.info({
            pluginName: pluginName,
            pluginCategory: pluginCategory
          }, 'plugin @{pluginName} successfully loaded (@{pluginCategory})');
          return _context2.a(3, 12);
        case 11:
          logger.error({
            pluginName: pluginName
          }, 'package not found, try to install @{pluginName} with a package manager');
          return _context2.a(3, 12);
        case 12:
          _i++;
          _context2.n = 1;
          break;
        case 13:
          debug('%o plugins found: %o', pluginCategory, plugins.length);
          return _context2.a(2, plugins);
      }
    }, _callee2, null, [[4, 8]]);
  }));
  return _asyncLoadPlugin.apply(this, arguments);
}
function executePlugin(plugin, pluginConfig, pluginOptions) {
  var legacyMergeConfigs = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : false;
  // this is a legacy support for plugins that are not using the new API
  if (legacyMergeConfigs) {
    debug('>>> plugin merge config enabled');
    var originalConfig = pluginOptions.config;
    pluginConfig = mergeConfig(originalConfig, pluginConfig);
  }
  if ((0, _utils.isES6)(plugin)) {
    debug('plugin is ES6');
    // @ts-expect-error no relevant for the code

    return new plugin["default"](pluginConfig, pluginOptions);
  } else {
    debug('plugin is commonJS');
    // @ts-expect-error improve this type
    return plugin(pluginConfig, pluginOptions);
  }
}
