// @verdaccio/file-locking/lib/utils.js
const require_runtime = require("./_virtual/_rolldown/runtime.js");
let lockfile = require("lockfile");
lockfile = require_runtime.__toESM(lockfile);
let node_fs = require("node:fs");
node_fs = require_runtime.__toESM(node_fs);
let node_path = require("node:path");
node_path = require_runtime.__toESM(node_path);
//#region src/utils.ts
var statDir = (name) => {
	return new Promise((resolve, reject) => {
		const dirPath = node_path.default.dirname(name);
		node_fs.default.stat(dirPath, function(err, stats) {
			if (err) return reject(err);
			else if (!stats.isDirectory()) return resolve(/* @__PURE__ */ new Error(`${node_path.default.dirname(name)} is not a directory`));
			else return resolve(null);
		});
	});
};
var statfile = (name) => {
	return new Promise((resolve, reject) => {
		node_fs.default.stat(name, function(err, stats) {
			if (err) return reject(err);
			else if (!stats.isFile()) return resolve(/* @__PURE__ */ new Error(`${node_path.default.dirname(name)} is not a file`));
			else return resolve(null);
		});
	});
};
var lockfile$1 = (name) => {
	return new Promise((resolve) => {
		const lockOpts = {
			wait: 1e3,
			pollPeriod: 100,
			stale: 300 * 1e3,
			retries: 100,
			retryWait: 100
		};
		const lockFileName = `${name}.lock`;
		lockfile.default.lock(lockFileName, lockOpts, () => {
			resolve(void 0);
		});
	});
};
//#endregion
exports.lockfile = lockfile$1;
exports.statDir = statDir;
exports.statfile = statfile;
