// @verdaccio/middleware/build/middlewares/request-options.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getRequestOptions = getRequestOptions;
var _debug = _interopRequireDefault(require("debug"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const debug = (0, _debug.default)('verdaccio:middleware:request-options');
function getRequestOptions(req) {
  const requestOptions = {
    // FIXME: decide if host should contain port or not
    //
    // Express 4:
    // - req.host is marked as deprecated since it does not include the port
    // - use req.headers['host'] instead, which includes the port
    // Express 5:
    // - req.host is fully supported and includes the port
    // - https://expressjs.com/en/api.html#req.host
    host: req.host,
    protocol: req.protocol,
    headers: req.headers,
    remoteAddress: req.socket.remoteAddress,
    byPassCache: req.query.write === 'true',
    username: req.remote_user?.name ?? undefined
  };
  debug('request options: %o', requestOptions);
  return requestOptions;
}
