// @verdaccio/package-filter/build/index.js
Object.defineProperties(exports, {
	__esModule: { value: true },
	[Symbol.toStringTag]: { value: "Module" }
});
//#region \0rolldown/runtime.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __copyProps = (to, from, except, desc) => {
	if (from && typeof from === "object" || typeof from === "function") for (var keys = __getOwnPropNames(from), i = 0, n = keys.length, key; i < n; i++) {
		key = keys[i];
		if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
			get: ((k) => from[k]).bind(null, key),
			enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
		});
	}
	return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
	value: mod,
	enumerable: true
}) : target, mod));
//#endregion
let debug = require("debug");
debug = __toESM(debug);
let _verdaccio_core = require("@verdaccio/core");
let semver = require("semver");
semver = __toESM(semver);
//#region src/config/parser.ts
var debug$6 = (0, debug.default)("verdaccio:plugin:package-filter:config");
function parseConfigRules(configRules) {
	const ruleMap = /* @__PURE__ */ new Map();
	for (const rule of configRules) {
		if ("scope" in rule && typeof rule.scope === "string") {
			if (!rule.scope.startsWith("@")) throw new TypeError(`Scope value must start with @, found: ${rule.scope}`);
			ruleMap.set(rule.scope, "scope");
			continue;
		}
		if ("package" in rule && !("versions" in rule)) {
			ruleMap.set(rule.package, "package");
			continue;
		}
		if ("package" in rule && "versions" in rule) {
			const previousConfig = ruleMap.get(rule.package) || { versions: [] };
			if (typeof previousConfig === "string") throw new Error(`Package ${rule.package} is already specified by another strict rule ${previousConfig}`);
			const range = new semver.Range(rule.versions);
			ruleMap.set(rule.package, {
				versions: [...previousConfig.versions, range],
				strategy: rule.strategy ?? "block"
			});
			continue;
		}
		throw new TypeError(`Could not parse rule ${JSON.stringify(rule, null, 4)}`);
	}
	return ruleMap;
}
function parseConfig(config) {
	debug$6("parsing config: %o", config);
	const blockMap = parseConfigRules(config.block ?? []);
	const allowMap = parseConfigRules(config.allow ?? []);
	debug$6("parsed %d block rules, %d allow rules", blockMap.size, allowMap.size);
	const dateThreshold = config.dateThreshold ? new Date(config.dateThreshold) : null;
	if (dateThreshold && isNaN(dateThreshold.getTime())) throw new TypeError(`Invalid date ${config.dateThreshold} was provided for dateThreshold`);
	const minAgeDays = config.minAgeDays ? Number(config.minAgeDays) : null;
	let minAgeMs = null;
	if (minAgeDays !== null) {
		if (isNaN(minAgeDays) || minAgeDays < 0) throw new TypeError(`Invalid number ${config.minAgeDays} was provided for minAgeDays`);
		minAgeMs = minAgeDays * 24 * 60 * 60 * 1e3;
	}
	return {
		dateThreshold,
		minAgeMs,
		blockRules: blockMap,
		allowRules: allowMap
	};
}
//#endregion
//#region src/filtering/types.ts
var MatchType = /* @__PURE__ */ function(MatchType) {
	MatchType["SCOPE"] = "scope";
	MatchType["PACKAGE"] = "package";
	MatchType["VERSIONS"] = "versions";
	return MatchType;
}({});
//#endregion
//#region src/filtering/matcher.ts
var debug$5 = (0, debug.default)("verdaccio:plugin:package-filter:filter");
/**
* Split a package name into name itself and scope.
*/
function splitName(name) {
	if (!name) return { name: "" };
	const parts = name.split("/");
	if (parts.length > 1) return {
		scope: parts[0],
		name: parts[1]
	};
	else return { name: parts[0] };
}
/**
* Try to find a rule that matches the package.
* If found, returns the rule and the matched package versions from the manifest.
*/
function matchRules(manifest, rules) {
	const { scope } = splitName(manifest.name);
	if (scope) {
		const rule = rules.get(scope);
		if (rule === "scope") {
			debug$5("scope match: %s matched rule for %s", manifest.name, scope);
			return {
				type: MatchType.SCOPE,
				rule,
				scope,
				versions: Object.keys(manifest.versions)
			};
		}
	}
	const rule = rules.get(manifest.name);
	if (!rule) return;
	if (rule === "package") {
		debug$5("package match: %s", manifest.name);
		return {
			type: MatchType.PACKAGE,
			rule,
			package: manifest.name,
			versions: Object.keys(manifest.versions)
		};
	}
	if (rule === "scope") throw new Error("Unexpected case - rule for package should never be \"scope\"");
	const versionRanges = rule.versions;
	if (versionRanges.length === 0) return;
	const matchedVersions = [];
	Object.keys(manifest.versions).forEach((version) => {
		versionRanges.forEach((versionRange) => {
			if ((0, semver.satisfies)(version, versionRange, {
				includePrerelease: true,
				loose: true
			})) matchedVersions.push(version);
		});
	});
	if (matchedVersions.length > 0) debug$5("version match: %s matched %d versions: %o", manifest.name, matchedVersions.length, matchedVersions);
	return {
		type: MatchType.VERSIONS,
		rule,
		versions: matchedVersions
	};
}
//#endregion
//#region src/filtering/packageVersion.ts
var debug$4 = (0, debug.default)("verdaccio:plugin:package-filter:filter");
/**
* Filter out all blocked package versions.
* If package or scope is blocked, then block all versions.
*
* TODO: consider adding a `prerelease` filter option to block/allow
* prerelease versions independently. Currently prereleases are matched by
* semver ranges with `includePrerelease: true`, but there's no dedicated toggle.
*
* Example config (not yet implemented):
*   block:
*     - package: 'foo'
*       prerelease: true    # block 1.0.0-beta.1, 2.0.0-rc.1, keep 1.0.0, 2.0.0
*     - package: 'bar'
*       prerelease: false   # block 1.0.0, 2.0.0, keep 1.0.0-beta.1
*
* Today the only workaround is using awkward ranges like ">=1.0.0-0 <1.0.0".
*/
function filterBlockedVersions(manifest, blockRules, allowRules, logger) {
	const allowMatch = matchRules(manifest, allowRules);
	if (allowMatch && (allowMatch.type === MatchType.SCOPE || allowMatch.type === MatchType.PACKAGE)) {
		logger.trace({ name: manifest.name }, "package @{name} is allow-listed, skipping block rules");
		return manifest;
	}
	const blockMatch = matchRules(manifest, blockRules);
	if (!blockMatch) return manifest;
	logger.trace({
		name: manifest.name,
		type: blockMatch.type
	}, "block rule matched for @{name} (type: @{type})");
	const whitelistedVersions = allowMatch ? allowMatch.versions : [];
	let blockRule = {
		versions: [new semver.Range("*")],
		strategy: "block"
	};
	if (blockMatch.type === MatchType.SCOPE) {
		if (whitelistedVersions.length === 0) {
			debug$4("blocking all versions of %s (scope %s blocked)", manifest.name, blockMatch.scope);
			logger.trace({
				name: manifest.name,
				scope: blockMatch.scope
			}, "all versions of @{name} blocked (scope @{scope})");
			return {
				...manifest,
				versions: {},
				readme: `All packages in scope ${blockMatch.scope} are blocked by rule`
			};
		}
	} else if (blockMatch.type === MatchType.PACKAGE) {
		if (whitelistedVersions.length === 0) {
			debug$4("blocking all versions of %s (package blocked)", manifest.name);
			logger.trace({ name: manifest.name }, "all versions of @{name} blocked (package rule)");
			return {
				...manifest,
				versions: {},
				readme: `All package versions are blocked by rule`
			};
		}
	} else blockRule = {
		...blockRule,
		...blockMatch.rule
	};
	const versionRanges = blockRule.versions;
	if (blockRule.strategy === "block") {
		const blockedVersions = blockMatch.versions.filter((v) => !whitelistedVersions.includes(v));
		for (const version of blockedVersions) delete manifest.versions[version];
		if (blockedVersions.length > 0) {
			debug$4("blocked %d versions of %s: %o", blockedVersions.length, manifest.name, blockedVersions);
			logger.trace({
				name: manifest.name,
				count: blockedVersions.length,
				versions: blockedVersions.join(", ")
			}, "@{count} versions of @{name} blocked: @{versions}");
			manifest.readme = (manifest.readme || "") + `\nSome versions(${blockedVersions.length}) of package are blocked by rules: ${versionRanges.map((range) => range.raw)}`;
		}
		return manifest;
	}
	const nonBlockedVersions = { ...manifest.versions };
	const newVersionsMapping = {};
	versionRanges.forEach((versionRange) => {
		const allVersions = Object.keys(nonBlockedVersions);
		let lastNonBlockedVersion = null;
		allVersions.forEach((version) => {
			if (!whitelistedVersions.includes(version) && (0, semver.satisfies)(version, versionRange, {
				includePrerelease: true,
				loose: true
			})) {
				delete nonBlockedVersions[version];
				newVersionsMapping[version] = lastNonBlockedVersion;
			} else lastNonBlockedVersion = version;
		});
	});
	debug$4("replacing versions for %s: %o", manifest.name, newVersionsMapping);
	const removedVersions = Object.entries(newVersionsMapping).filter(([_, replace]) => replace === null);
	const replacedVersions = Object.entries(newVersionsMapping).filter(([_, replace]) => replace !== null);
	removedVersions.forEach(([version]) => {
		debug$4("no version to replace %s in %s", version, manifest.name);
		logger.trace({
			name: manifest.name,
			version
		}, "version @{version} of @{name} removed (no replacement available)");
		delete manifest.versions[version];
	});
	replacedVersions.forEach(([version, replaceVersion]) => {
		logger.trace({
			name: manifest.name,
			version,
			replaceVersion
		}, "version @{version} of @{name} replaced with @{replaceVersion}");
		manifest.versions[version] = {
			...manifest.versions[replaceVersion],
			version
		};
	});
	if (removedVersions.length > 0) manifest.readme += `\nSome versions of package could not be replaced and thus are fully blocked (${removedVersions.length}): ${removedVersions.map((a) => a[0])}`;
	if (replacedVersions.length > 0) manifest.readme += `\nSome versions of package are replaced by other(${replacedVersions.length}): ${replacedVersions.map((a) => `${a[0]} => ${a[1]}`)}`;
	return manifest;
}
//#endregion
//#region src/filtering/publishDate.ts
var debug$3 = (0, debug.default)("verdaccio:plugin:package-filter:filter");
/**
* Filter out all package versions that were published after dateThreshold.
*/
function filterVersionsByPublishDate(manifest, dateThreshold, allowRules) {
	const allowMatch = matchRules(manifest, allowRules);
	if (allowMatch && (allowMatch.type === MatchType.SCOPE || allowMatch.type === MatchType.PACKAGE)) return manifest;
	const { versions, time, name } = manifest;
	if (!time) throw new TypeError(`Time of publication was not provided for package ${name}`);
	const whitelistedVersions = allowMatch ? allowMatch.versions : [];
	const clearVersions = [];
	Object.keys(versions).forEach((version) => {
		if (whitelistedVersions.includes(version)) return;
		const publishTime = time[version];
		if (!publishTime) throw new TypeError(`Time of publication was not provided for package ${name}, version ${version}`);
		if (new Date(publishTime) > dateThreshold) clearVersions.push(version);
	});
	clearVersions.forEach((version) => {
		delete manifest.versions[version];
	});
	if (clearVersions.length > 0) debug$3("date filter removed %d versions from %s: %o", clearVersions.length, manifest.name, clearVersions);
	return manifest;
}
//#endregion
//#region src/utils/jsonUtils.ts
function jsonLogReplacer(_key, value) {
	if (value instanceof Map) return Object.fromEntries(value);
	if (value instanceof semver.Range) return value.range;
	return value;
}
//#endregion
//#region src/utils/manifestUtils.ts
var debug$2 = (0, debug.default)("verdaccio:plugin:package-filter:manifest");
/**
* Delete `dist-tags` entries corresponding to missing versions.
*/
function cleanupTags(manifest) {
	const distTags = manifest[_verdaccio_core.DIST_TAGS];
	Object.entries(distTags).forEach(([tag, tagVersion]) => {
		if (!manifest.versions[tagVersion]) {
			debug$2("removing orphaned dist-tag %s -> %s from %s", tag, tagVersion, manifest.name);
			delete distTags[tag];
		}
	});
}
/**
* Delete `time` entries corresponding to missing versions.
*/
function cleanupTime(manifest) {
	const time = manifest.time;
	if (!time) return;
	Object.keys(time).forEach((version) => {
		if (!manifest.versions[version]) delete time[version];
	});
}
/**
* Get the latest version from a list of versions,
* ordered by time of their publication stored in the manifest.
*/
function getLatestVersion(manifest, versions) {
	const time = manifest.time;
	if (!time) return versions.sort(semver.default.rcompare)[0];
	const timedVersions = versions.map((v) => ({
		version: v,
		time: time[v]
	})).filter((v) => v.time);
	if (timedVersions.length === 0) return;
	return timedVersions.sort((a, b) => new Date(b.time).getTime() - new Date(a.time).getTime())[0].version;
}
/**
* Set the latest tag if dist-tags/latest is missing.
* The last stable version available is used when possible.
* Otherwise, it uses the latest version not found in dist-tags.
*/
function setupLatestTag(manifest) {
	const distTags = manifest[_verdaccio_core.DIST_TAGS];
	if (distTags.latest) return;
	const versions = Object.keys(manifest.versions);
	if (versions.length === 0) return;
	const distTagsVersions = Object.values(distTags);
	const untaggedVersions = versions.filter((v) => semver.default.valid(v) && !distTagsVersions.includes(v));
	if (untaggedVersions.length === 0) return;
	const latestStableVersion = getLatestVersion(manifest, untaggedVersions.filter((v) => !semver.default.prerelease(v)));
	if (latestStableVersion) {
		debug$2("reassigned latest tag to stable version %s for %s", latestStableVersion, manifest.name);
		distTags.latest = latestStableVersion;
		return;
	}
	const latestVersion = getLatestVersion(manifest, untaggedVersions);
	if (!latestVersion) return;
	debug$2("reassigned latest tag to pre-release version %s for %s", latestVersion, manifest.name);
	distTags.latest = latestVersion;
}
/**
* Set the created and modified times.
*/
function setupCreatedAndModified(manifest) {
	const time = manifest.time;
	if (!time) return;
	const times = Object.values(time);
	if (times.length === 0) return;
	const sortedTimes = times.sort((a, b) => new Date(a).getTime() - new Date(b).getTime());
	time.created = sortedTimes[0];
	time.modified = sortedTimes[sortedTimes.length - 1];
}
/**
* Remove `_distfiles` entries which are not used by any version.
*/
function cleanupDistFiles(manifest) {
	const distFiles = manifest._distfiles;
	const activeTarballs = new Set(Object.values(manifest.versions).map((v) => v.dist?.tarball).filter((tarball) => typeof tarball === "string"));
	Object.keys(distFiles).forEach((key) => {
		if (!activeTarballs.has(distFiles[key].url)) delete distFiles[key];
	});
}
/**
* Creates a copy of a manifest suitable for safe, localized mutation.
*
* The returned object is shallow-cloned, except for `versions`, `dist-tags`,
* `time`, and `_distfiles`, which are cloned as independent maps so they can be
* filtered or modified without affecting the original manifest.
*/
function getManifestClone(manifest) {
	return {
		...manifest,
		versions: { ...manifest.versions },
		[_verdaccio_core.DIST_TAGS]: { ...manifest[_verdaccio_core.DIST_TAGS] },
		time: { ...manifest.time },
		_distfiles: { ...manifest._distfiles }
	};
}
//#endregion
//#region src/packageFilter.ts
var debug$1 = (0, debug.default)("verdaccio:plugin:package-filter");
var PackageFilterPlugin = class extends _verdaccio_core.pluginUtils.Plugin {
	config;
	parsedConfig;
	logger;
	constructor(config, options) {
		super(config, options);
		this.config = config ?? {};
		this.logger = options.logger;
		this.parsedConfig = parseConfig(this.config);
		debug$1("plugin loaded: block rules: %d, allow rules: %d", this.parsedConfig.blockRules.size, this.parsedConfig.allowRules.size);
		this.logger.debug({ config: JSON.stringify(this.parsedConfig, jsonLogReplacer) }, "package-filter loaded with config: @{config}");
		this.logger.trace({
			blockRules: this.parsedConfig.blockRules.size,
			allowRules: this.parsedConfig.allowRules.size
		}, "package-filter plugin initialized: @{blockRules} block rules, @{allowRules} allow rules");
		if (this.parsedConfig.dateThreshold) {
			debug$1("date threshold: %s", this.parsedConfig.dateThreshold.toISOString());
			this.logger.trace({ dateThreshold: this.parsedConfig.dateThreshold.toISOString() }, "package-filter date threshold: @{dateThreshold}");
		}
		if (this.parsedConfig.minAgeMs) {
			const minAgeDays = this.parsedConfig.minAgeMs / (1440 * 60 * 1e3);
			debug$1("min age: %d days", minAgeDays);
			this.logger.trace({ minAgeDays }, "package-filter min age: @{minAgeDays} days");
		}
	}
	async filter_metadata(manifest) {
		const { dateThreshold, minAgeMs, blockRules, allowRules } = this.parsedConfig;
		const versionCount = Object.keys(manifest.versions ?? {}).length;
		debug$1("filtering manifest for %s (%d versions)", manifest.name, versionCount);
		this.logger.trace({
			name: manifest.name,
			versionCount
		}, "package-filter processing @{name} (@{versionCount} versions)");
		let newManifest = getManifestClone(manifest);
		if (blockRules.size > 0) newManifest = filterBlockedVersions(newManifest, blockRules, allowRules, this.logger);
		let earliestDateThreshold = null;
		if (minAgeMs) earliestDateThreshold = new Date(Date.now() - minAgeMs);
		if (dateThreshold && (!earliestDateThreshold || dateThreshold < earliestDateThreshold)) earliestDateThreshold = dateThreshold;
		if (earliestDateThreshold) {
			debug$1("applying date filter for %s, threshold: %s", manifest.name, earliestDateThreshold.toISOString());
			this.logger.trace({
				name: manifest.name,
				threshold: earliestDateThreshold.toISOString()
			}, "applying date filter for @{name}, cutoff: @{threshold}");
			newManifest = filterVersionsByPublishDate(newManifest, earliestDateThreshold, allowRules);
		}
		cleanupTags(newManifest);
		setupLatestTag(newManifest);
		cleanupTime(newManifest);
		setupCreatedAndModified(newManifest);
		cleanupDistFiles(newManifest);
		const filteredCount = Object.keys(newManifest.versions).length;
		const removedCount = versionCount - filteredCount;
		if (filteredCount !== versionCount) {
			debug$1("filtered %s: %d -> %d versions (%d removed)", manifest.name, versionCount, filteredCount, removedCount);
			this.logger.trace({
				name: manifest.name,
				before: versionCount,
				after: filteredCount,
				removed: removedCount
			}, "package-filter @{name}: @{before} -> @{after} versions (@{removed} removed)");
		} else debug$1("no versions filtered for %s", manifest.name);
		return newManifest;
	}
};
//#endregion
//#region src/index.ts
var src_default = PackageFilterPlugin;
//#endregion
exports.PackageFilterPlugin = PackageFilterPlugin;
exports.default = src_default;
