// @verdaccio/hooks/build/notify.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.compileTemplate = compileTemplate;
exports.handleNotify = handleNotify;
exports.notify = notify;
exports.sendNotification = sendNotification;
var _debug = _interopRequireDefault(require("debug"));
var _handlebars = _interopRequireDefault(require("handlebars"));
var _logger = require("@verdaccio/logger");
var _notifyRequest = require("./notify-request");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { "default": e }; }
function _slicedToArray(r, e) { return _arrayWithHoles(r) || _iterableToArrayLimit(r, e) || _unsupportedIterableToArray(r, e) || _nonIterableRest(); }
function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _unsupportedIterableToArray(r, a) { if (r) { if ("string" == typeof r) return _arrayLikeToArray(r, a); var t = {}.toString.call(r).slice(8, -1); return "Object" === t && r.constructor && (t = r.constructor.name), "Map" === t || "Set" === t ? Array.from(r) : "Arguments" === t || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(t) ? _arrayLikeToArray(r, a) : void 0; } }
function _arrayLikeToArray(r, a) { (null == a || a > r.length) && (a = r.length); for (var e = 0, n = Array(a); e < a; e++) n[e] = r[e]; return n; }
function _iterableToArrayLimit(r, l) { var t = null == r ? null : "undefined" != typeof Symbol && r[Symbol.iterator] || r["@@iterator"]; if (null != t) { var e, n, i, u, a = [], f = !0, o = !1; try { if (i = (t = t.call(r)).next, 0 === l) { if (Object(t) !== t) return; f = !1; } else for (; !(f = (e = i.call(t)).done) && (a.push(e.value), a.length !== l); f = !0); } catch (r) { o = !0, n = r; } finally { try { if (!f && null != t["return"] && (u = t["return"](), Object(u) !== u)) return; } finally { if (o) throw n; } } return a; } }
function _arrayWithHoles(r) { if (Array.isArray(r)) return r; }
function _regenerator() { /*! regenerator-runtime -- Copyright (c) 2014-present, Facebook, Inc. -- license (MIT): https://github.com/babel/babel/blob/main/packages/babel-helpers/LICENSE */ var e, t, r = "function" == typeof Symbol ? Symbol : {}, n = r.iterator || "@@iterator", o = r.toStringTag || "@@toStringTag"; function i(r, n, o, i) { var c = n && n.prototype instanceof Generator ? n : Generator, u = Object.create(c.prototype); return _regeneratorDefine2(u, "_invoke", function (r, n, o) { var i, c, u, f = 0, p = o || [], y = !1, G = { p: 0, n: 0, v: e, a: d, f: d.bind(e, 4), d: function d(t, r) { return i = t, c = 0, u = e, G.n = r, a; } }; function d(r, n) { for (c = r, u = n, t = 0; !y && f && !o && t < p.length; t++) { var o, i = p[t], d = G.p, l = i[2]; r > 3 ? (o = l === n) && (u = i[(c = i[4]) ? 5 : (c = 3, 3)], i[4] = i[5] = e) : i[0] <= d && ((o = r < 2 && d < i[1]) ? (c = 0, G.v = n, G.n = i[1]) : d < l && (o = r < 3 || i[0] > n || n > l) && (i[4] = r, i[5] = n, G.n = l, c = 0)); } if (o || r > 1) return a; throw y = !0, n; } return function (o, p, l) { if (f > 1) throw TypeError("Generator is already running"); for (y && 1 === p && d(p, l), c = p, u = l; (t = c < 2 ? e : u) || !y;) { i || (c ? c < 3 ? (c > 1 && (G.n = -1), d(c, u)) : G.n = u : G.v = u); try { if (f = 2, i) { if (c || (o = "next"), t = i[o]) { if (!(t = t.call(i, u))) throw TypeError("iterator result is not an object"); if (!t.done) return t; u = t.value, c < 2 && (c = 0); } else 1 === c && (t = i["return"]) && t.call(i), c < 2 && (u = TypeError("The iterator does not provide a '" + o + "' method"), c = 1); i = e; } else if ((t = (y = G.n < 0) ? u : r.call(n, G)) !== a) break; } catch (t) { i = e, c = 1, u = t; } finally { f = 1; } } return { value: t, done: y }; }; }(r, o, i), !0), u; } var a = {}; function Generator() {} function GeneratorFunction() {} function GeneratorFunctionPrototype() {} t = Object.getPrototypeOf; var c = [][n] ? t(t([][n]())) : (_regeneratorDefine2(t = {}, n, function () { return this; }), t), u = GeneratorFunctionPrototype.prototype = Generator.prototype = Object.create(c); function f(e) { return Object.setPrototypeOf ? Object.setPrototypeOf(e, GeneratorFunctionPrototype) : (e.__proto__ = GeneratorFunctionPrototype, _regeneratorDefine2(e, o, "GeneratorFunction")), e.prototype = Object.create(u), e; } return GeneratorFunction.prototype = GeneratorFunctionPrototype, _regeneratorDefine2(u, "constructor", GeneratorFunctionPrototype), _regeneratorDefine2(GeneratorFunctionPrototype, "constructor", GeneratorFunction), GeneratorFunction.displayName = "GeneratorFunction", _regeneratorDefine2(GeneratorFunctionPrototype, o, "GeneratorFunction"), _regeneratorDefine2(u), _regeneratorDefine2(u, o, "Generator"), _regeneratorDefine2(u, n, function () { return this; }), _regeneratorDefine2(u, "toString", function () { return "[object Generator]"; }), (_regenerator = function _regenerator() { return { w: i, m: f }; })(); }
function _regeneratorDefine2(e, r, n, t) { var i = Object.defineProperty; try { i({}, "", {}); } catch (e) { i = 0; } _regeneratorDefine2 = function _regeneratorDefine(e, r, n, t) { function o(r, n) { _regeneratorDefine2(e, r, function (e) { return this._invoke(r, n, e); }); } r ? i ? i(e, r, { value: n, enumerable: !t, configurable: !t, writable: !t }) : e[r] = n : (o("next", 0), o("throw", 1), o("return", 2)); }, _regeneratorDefine2(e, r, n, t); }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(e, r, t) { return (r = _toPropertyKey(r)) in e ? Object.defineProperty(e, r, { value: t, enumerable: !0, configurable: !0, writable: !0 }) : e[r] = t, e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == _typeof(i) ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != _typeof(t) || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != _typeof(i)) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
function asyncGeneratorStep(n, t, e, r, o, a, c) { try { var i = n[a](c), u = i.value; } catch (n) { return void e(n); } i.done ? t(u) : Promise.resolve(u).then(r, o); }
function _asyncToGenerator(n) { return function () { var t = this, e = arguments; return new Promise(function (r, o) { var a = n.apply(t, e); function _next(n) { asyncGeneratorStep(a, r, o, _next, _throw, "next", n); } function _throw(n) { asyncGeneratorStep(a, r, o, _next, _throw, "throw", n); } _next(void 0); }); }; }
var debug = (0, _debug["default"])('verdaccio:hooks');
function compileTemplate(content, metadata) {
  // FUTURE: multiple handlers
  return new Promise(function (resolve, reject) {
    var handler;
    try {
      if (!handler) {
        debug('compile default template handler %o', content);
        var template = _handlebars["default"].compile(content);
        return resolve(template(metadata));
      }
    } catch (error) {
      debug('error  template handler %o', error === null || error === void 0 ? void 0 : error.message);
      _logger.logger.error({
        error: error.message
      }, 'notification template compilation has failed: @{error}');
      reject(error);
    }
  });
}
function handleNotify(_x, _x2, _x3, _x4) {
  return _handleNotify.apply(this, arguments);
}
function _handleNotify() {
  _handleNotify = _asyncToGenerator(/*#__PURE__*/_regenerator().m(function _callee(metadata, notifyEntry, remoteUser, publishedPackage) {
    var _metadata, _metadata2;
    var regex, _notifyEntry$packageP, content, options, header, _t;
    return _regenerator().w(function (_context) {
      while (1) switch (_context.p = _context.n) {
        case 0:
          _context.p = 0;
          if (!(metadata.name && notifyEntry.packagePattern)) {
            _context.n = 2;
            break;
          }
          debug('checking package pattern %o', notifyEntry.packagePattern);
          debug('checking package pattern flags %o', notifyEntry.packagePatternFlags);
          regex = new RegExp(notifyEntry.packagePattern, (_notifyEntry$packageP = notifyEntry.packagePatternFlags) !== null && _notifyEntry$packageP !== void 0 ? _notifyEntry$packageP : '');
          if (regex.test(metadata.name)) {
            _context.n = 1;
            break;
          }
          _logger.logger.debug('Notification exclude does not match %o', metadata.name);
          return _context.a(2, false);
        case 1:
          debug('package pattern matched %o', metadata.name);
        case 2:
          _context.n = 4;
          break;
        case 3:
          _context.p = 3;
          _t = _context.v;
          _logger.logger.error('Invalid regex pattern or flags: %o', _t === null || _t === void 0 ? void 0 : _t.message);
          return _context.a(2, false);
        case 4:
          if (!(typeof ((_metadata = metadata) === null || _metadata === void 0 ? void 0 : _metadata.publisher) === 'undefined' || ((_metadata2 = metadata) === null || _metadata2 === void 0 ? void 0 : _metadata2.publisher) === null)) {
            _context.n = 6;
            break;
          }
          // @ts-ignore
          metadata = _objectSpread(_objectSpread({}, metadata), {}, {
            publishedPackage: publishedPackage,
            publisher: {
              name: remoteUser.name
            }
          });
          debug('template metadata %o', metadata);
          _context.n = 5;
          return compileTemplate(notifyEntry.content, metadata);
        case 5:
          content = _context.v;
        case 6:
          options = {
            body: content
          }; // provides fallback support, it's accept an Object {} and Array of {}
          if (notifyEntry.headers && Array.isArray(notifyEntry.headers)) {
            header = {}; // FIXME: we can simplify this
            notifyEntry.headers.map(function (item) {
              if (Object.is(item, item)) {
                for (var key in item) {
                  /* eslint no-prototype-builtins: 0 */
                  if (item.hasOwnProperty(key)) {
                    header[key] = item[key];
                  }
                }
              }
            });
            options.headers = header;
          } else if (Object.is(notifyEntry.headers, notifyEntry.headers)) {
            options.headers = notifyEntry.headers;
          }
          if (notifyEntry.endpoint) {
            _context.n = 7;
            break;
          }
          debug('error due endpoint is missing');
          throw new Error('missing parameter');
        case 7:
          return _context.a(2, (0, _notifyRequest.notifyRequest)(notifyEntry.endpoint, _objectSpread({
            method: notifyEntry.method
          }, options)));
      }
    }, _callee, null, [[0, 3]]);
  }));
  return _handleNotify.apply(this, arguments);
}
function sendNotification(metadata, notify, remoteUser, publishedPackage) {
  return handleNotify(metadata, notify, remoteUser, publishedPackage);
}
function isHasNotification(value) {
  return value && _typeof(value) === 'object' && 'endpoint' in value && 'content' in value;
}
function notify(_x5, _x6, _x7, _x8) {
  return _notify.apply(this, arguments);
}
function _notify() {
  _notify = _asyncToGenerator(/*#__PURE__*/_regenerator().m(function _callee2(metadata, config, remoteUser, publishedPackage) {
    var notification, isSingle, response, notificationEntries, results, _t2, _t3;
    return _regenerator().w(function (_context2) {
      while (1) switch (_context2.p = _context2.n) {
        case 0:
          debug('init send notification');
          notification = config === null || config === void 0 ? void 0 : config.notify;
          if (notification) {
            _context2.n = 1;
            break;
          }
          debug('no notify configuration detected');
          return _context2.a(2, [false]);
        case 1:
          isSingle = isHasNotification(notification);
          debug('is single notify %o', isSingle);
          if (!isSingle) {
            _context2.n = 5;
            break;
          }
          debug('send single notification');
          _context2.p = 2;
          _context2.n = 3;
          return sendNotification(metadata, config.notify, remoteUser, publishedPackage);
        case 3:
          response = _context2.v;
          return _context2.a(2, [response]);
        case 4:
          _context2.p = 4;
          _t2 = _context2.v;
          debug('error on sending single notification');
          return _context2.a(2, [false]);
        case 5:
          _context2.p = 5;
          debug('send multiples notification');
          notificationEntries = Object.entries(notification).filter(function (_ref) {
            var _ref2 = _slicedToArray(_ref, 2),
              item = _ref2[1];
            return isHasNotification(item);
          });
          debug('valid notification entries %o', notificationEntries.length);
          if (!(notificationEntries.length === 0)) {
            _context2.n = 6;
            break;
          }
          return _context2.a(2, Object.keys(notification).map(function () {
            return false;
          }));
        case 6:
          debug('sending %o notifications', notificationEntries.length);
          _context2.n = 7;
          return Promise.allSettled(notificationEntries.map(function (_ref3) {
            var _ref4 = _slicedToArray(_ref3, 2),
              item = _ref4[1];
            return sendNotification(metadata, item, remoteUser, publishedPackage);
          }))["catch"](function (error) {
            _logger.logger.error({
              error: error
            }, 'notify request has failed: @error');
          });
        case 7:
          results = _context2.v;
          if (results) {
            _context2.n = 8;
            break;
          }
          return _context2.a(2, []);
        case 8:
          return _context2.a(2, results.map(function (result) {
            return result.status === 'fulfilled' ? result.value : false;
          }));
        case 9:
          _context2.p = 9;
          _t3 = _context2.v;
          debug('error on sending multiple notification %o', _t3);
          return _context2.a(2, Object.keys(notification).map(function () {
            return false;
          }));
        case 10:
          return _context2.a(2);
      }
    }, _callee2, null, [[5, 9], [2, 4]]);
  }));
  return _notify.apply(this, arguments);
}
