// @verdaccio/hooks/build/notify-request.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.notifyRequest = notifyRequest;
exports.verifyMethod = verifyMethod;
var _debug = _interopRequireDefault(require("debug"));
var _gotCjs = _interopRequireDefault(require("got-cjs"));
var _core = require("@verdaccio/core");
var _logger = require("@verdaccio/logger");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { "default": e }; }
function _regenerator() { /*! regenerator-runtime -- Copyright (c) 2014-present, Facebook, Inc. -- license (MIT): https://github.com/babel/babel/blob/main/packages/babel-helpers/LICENSE */ var e, t, r = "function" == typeof Symbol ? Symbol : {}, n = r.iterator || "@@iterator", o = r.toStringTag || "@@toStringTag"; function i(r, n, o, i) { var c = n && n.prototype instanceof Generator ? n : Generator, u = Object.create(c.prototype); return _regeneratorDefine2(u, "_invoke", function (r, n, o) { var i, c, u, f = 0, p = o || [], y = !1, G = { p: 0, n: 0, v: e, a: d, f: d.bind(e, 4), d: function d(t, r) { return i = t, c = 0, u = e, G.n = r, a; } }; function d(r, n) { for (c = r, u = n, t = 0; !y && f && !o && t < p.length; t++) { var o, i = p[t], d = G.p, l = i[2]; r > 3 ? (o = l === n) && (u = i[(c = i[4]) ? 5 : (c = 3, 3)], i[4] = i[5] = e) : i[0] <= d && ((o = r < 2 && d < i[1]) ? (c = 0, G.v = n, G.n = i[1]) : d < l && (o = r < 3 || i[0] > n || n > l) && (i[4] = r, i[5] = n, G.n = l, c = 0)); } if (o || r > 1) return a; throw y = !0, n; } return function (o, p, l) { if (f > 1) throw TypeError("Generator is already running"); for (y && 1 === p && d(p, l), c = p, u = l; (t = c < 2 ? e : u) || !y;) { i || (c ? c < 3 ? (c > 1 && (G.n = -1), d(c, u)) : G.n = u : G.v = u); try { if (f = 2, i) { if (c || (o = "next"), t = i[o]) { if (!(t = t.call(i, u))) throw TypeError("iterator result is not an object"); if (!t.done) return t; u = t.value, c < 2 && (c = 0); } else 1 === c && (t = i["return"]) && t.call(i), c < 2 && (u = TypeError("The iterator does not provide a '" + o + "' method"), c = 1); i = e; } else if ((t = (y = G.n < 0) ? u : r.call(n, G)) !== a) break; } catch (t) { i = e, c = 1, u = t; } finally { f = 1; } } return { value: t, done: y }; }; }(r, o, i), !0), u; } var a = {}; function Generator() {} function GeneratorFunction() {} function GeneratorFunctionPrototype() {} t = Object.getPrototypeOf; var c = [][n] ? t(t([][n]())) : (_regeneratorDefine2(t = {}, n, function () { return this; }), t), u = GeneratorFunctionPrototype.prototype = Generator.prototype = Object.create(c); function f(e) { return Object.setPrototypeOf ? Object.setPrototypeOf(e, GeneratorFunctionPrototype) : (e.__proto__ = GeneratorFunctionPrototype, _regeneratorDefine2(e, o, "GeneratorFunction")), e.prototype = Object.create(u), e; } return GeneratorFunction.prototype = GeneratorFunctionPrototype, _regeneratorDefine2(u, "constructor", GeneratorFunctionPrototype), _regeneratorDefine2(GeneratorFunctionPrototype, "constructor", GeneratorFunction), GeneratorFunction.displayName = "GeneratorFunction", _regeneratorDefine2(GeneratorFunctionPrototype, o, "GeneratorFunction"), _regeneratorDefine2(u), _regeneratorDefine2(u, o, "Generator"), _regeneratorDefine2(u, n, function () { return this; }), _regeneratorDefine2(u, "toString", function () { return "[object Generator]"; }), (_regenerator = function _regenerator() { return { w: i, m: f }; })(); }
function _regeneratorDefine2(e, r, n, t) { var i = Object.defineProperty; try { i({}, "", {}); } catch (e) { i = 0; } _regeneratorDefine2 = function _regeneratorDefine(e, r, n, t) { function o(r, n) { _regeneratorDefine2(e, r, function (e) { return this._invoke(r, n, e); }); } r ? i ? i(e, r, { value: n, enumerable: !t, configurable: !t, writable: !t }) : e[r] = n : (o("next", 0), o("throw", 1), o("return", 2)); }, _regeneratorDefine2(e, r, n, t); }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(e, r, t) { return (r = _toPropertyKey(r)) in e ? Object.defineProperty(e, r, { value: t, enumerable: !0, configurable: !0, writable: !0 }) : e[r] = t, e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == _typeof(i) ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != _typeof(t) || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != _typeof(i)) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
function asyncGeneratorStep(n, t, e, r, o, a, c) { try { var i = n[a](c), u = i.value; } catch (n) { return void e(n); } i.done ? t(u) : Promise.resolve(u).then(r, o); }
function _asyncToGenerator(n) { return function () { var t = this, e = arguments; return new Promise(function (r, o) { var a = n.apply(t, e); function _next(n) { asyncGeneratorStep(a, r, o, _next, _throw, "next", n); } function _throw(n) { asyncGeneratorStep(a, r, o, _next, _throw, "throw", n); } _next(void 0); }); }; }
var debug = (0, _debug["default"])('verdaccio:hooks:request');
function verifyMethod(value) {
  var valid = ['GET', 'POST', 'PUT'];
  if (typeof value === 'string') {
    var upper = value.toUpperCase();
    if (valid.includes(upper)) {
      return upper;
    } else {
      _logger.logger.warn({
        method: value
      }, 'The notification method @{method} is not valid, using default POST method');
    }
  }
  return 'POST';
}
var baseHeaders = {
  'Content-Type': 'application/json'
};
function notifyRequest(_x, _x2) {
  return _notifyRequest.apply(this, arguments);
}
function _notifyRequest() {
  _notifyRequest = _asyncToGenerator(/*#__PURE__*/_regenerator().m(function _callee(url, options) {
    var response, method, userHeaders, headers, requestOptions, finalUrl, urlObj, params, body, _t;
    return _regenerator().w(function (_context) {
      while (1) switch (_context.p = _context.n) {
        case 0:
          _context.p = 0;
          method = verifyMethod(options.method);
          debug('uri %o', url);
          debug('headers %o', options.headers);
          debug('method %o', method);
          userHeaders = {};
          if (options.headers) {
            if (typeof options.headers === 'string') {
              userHeaders = JSON.parse(options.headers);
            } else if (_typeof(options.headers) === 'object') {
              userHeaders = options.headers;
            }
          }
          headers = _objectSpread(_objectSpread({}, baseHeaders), userHeaders);
          requestOptions = {
            method: method,
            headers: headers
          };
          finalUrl = url;
          if (!(method === 'GET')) {
            _context.n = 1;
            break;
          }
          debug('using GET with query params');
          urlObj = new URL(url);
          params = new URLSearchParams(options.body);
          params.set('body', options.body);
          urlObj.search = params.toString();
          finalUrl = urlObj.toString();
          debug('final url with search params %o', finalUrl);
          _context.n = 3;
          break;
        case 1:
          if (!(options.body !== undefined)) {
            _context.n = 2;
            break;
          }
          requestOptions.body = options.body;
          _context.n = 3;
          break;
        case 2:
          throw new Error('Notification body is undefined');
        case 3:
          _context.n = 4;
          return (0, _gotCjs["default"])(finalUrl, _objectSpread(_objectSpread({}, requestOptions), {}, {
            responseType: 'json'
          }));
        case 4:
          response = _context.v;
          _context.n = 5;
          return response.body;
        case 5:
          body = _context.v;
          debug('response.status %o', body.statusCode);
          if (!(body.statusCode >= _core.HTTP_STATUS.BAD_REQUEST)) {
            _context.n = 6;
            break;
          }
          throw new Error(body);
        case 6:
          _logger.logger.info({
            content: options.body
          }, 'The notification @{content} has been successfully dispatched');
          return _context.a(2, true);
        case 7:
          _context.p = 7;
          _t = _context.v;
          debug('request error %o:', _t === null || _t === void 0 ? void 0 : _t.message);
          _logger.logger.error({
            errorMessage: _t === null || _t === void 0 ? void 0 : _t.message
          }, 'notify service has thrown an error: @{errorMessage}');
          return _context.a(2, false);
      }
    }, _callee, null, [[0, 7]]);
  }));
  return _notifyRequest.apply(this, arguments);
}
